--- gcc-3.0.3/gcc/gcc.c	Sat Dec 22 01:40:45 2001
+++ gcc-3.0.3/gcc/gcc.c.new	Sun Dec 23 18:43:36 2001
@@ -1160,16 +1160,16 @@
 /* Supply defaults for the standard prefixes.  */
 
 #ifndef STANDARD_EXEC_PREFIX
-#define STANDARD_EXEC_PREFIX "/usr/local/lib/gcc-lib/"
+#define STANDARD_EXEC_PREFIX "/usr/lib/gcc-lib/"
 #endif
 #ifndef STANDARD_STARTFILE_PREFIX
-#define STANDARD_STARTFILE_PREFIX "/usr/local/lib/"
+#define STANDARD_STARTFILE_PREFIX "/usr/lib/"
 #endif
 #ifndef TOOLDIR_BASE_PREFIX
-#define TOOLDIR_BASE_PREFIX "/usr/local/"
+#define TOOLDIR_BASE_PREFIX "/usr/"
 #endif
 #ifndef STANDARD_BINDIR_PREFIX
-#define STANDARD_BINDIR_PREFIX "/usr/local/bin"
+#define STANDARD_BINDIR_PREFIX "/usr/bin"
 #endif
 
 static const char *standard_exec_prefix = STANDARD_EXEC_PREFIX;
