diff -ur gcc-20020304/gcc/ada/Makefile.in gcc-20020304-/gcc/ada/Makefile.in
--- gcc-20020304/gcc/ada/Makefile.in	Fri Jan  4 19:29:52 2002
+++ gcc-20020304-/gcc/ada/Makefile.in	Thu Mar  7 21:40:11 2002
@@ -1165,7 +1165,7 @@
   system.ads<5lsystem.ads
 
   MLIB_TGT=5lml-tgt
-  MISCLIB=-laddr2line -lbfd
+  MISCLIB=
   THREADSLIB=-lpthread
   GNATLIB_SHARED=gnatlib-shared-dual
   GMEM_LIB=gmemlib
diff -ur gcc-20020304/gcc/ada/adaint.c gcc-20020304-/gcc/ada/adaint.c
--- gcc-20020304/gcc/ada/adaint.c	Tue Feb 12 17:32:00 2002
+++ gcc-20020304-/gcc/ada/adaint.c	Thu Mar  7 21:38:32 2002
@@ -2214,7 +2214,7 @@
 }
 #endif
 
-#if defined (CROSS_COMPILE)  \
+#if 1 || defined (CROSS_COMPILE)  \
   || (! (defined (sparc) && defined (sun) && defined (__SVR4)) \
       && ! defined (linux) \
       && ! defined (sgi) \
