--- gcc-4.0-20040919/gcc/ada/Make-lang.in.orig	2004-09-01 12:46:47.000000000 +0200
+++ gcc-4.0-20040919/gcc/ada/Make-lang.in	2004-09-26 12:21:15.135188144 +0200
@@ -378,7 +378,7 @@
 ada/doctools/xgnatugn$(build_exeext): ada/xgnatugn.adb
 	-$(MKDIR) ada/doctools
 	$(CP) $^ ada/doctools
-	cd ada/doctools && gnatmake -q xgnatugn
+	cd ada/doctools && gnatmake -q xgnatugn -a0../rts
 
 # Note that gnat_ugn_unw.texi does not depend on xgnatugn 
 # being built so we can distribute a pregenerated gnat_ugn_unw.info
