--- gcc-4.1-20050624_1105/gcc/tree-ssa-operands.c.orig	2005-06-24 11:01:40.000000000 +0200
+++ gcc-4.1-20050624_1105/gcc/tree-ssa-operands.c	2005-06-24 15:35:28.000000000 +0200
@@ -1347,7 +1347,7 @@
 	
 	if (code == COMPONENT_REF)
 	  {
-	    if (TREE_THIS_VOLATILE (TREE_OPERAND (expr, 1)))
+	    if (s_ann && TREE_THIS_VOLATILE (TREE_OPERAND (expr, 1)))
 	      s_ann->has_volatile_ops = true; 
 	    get_expr_operands (stmt, &TREE_OPERAND (expr, 2), opf_none);
 	  }
