--- trunk/libjava/classpath/configure.ac.orig	2006-07-02 15:32:04.000000000 +0200
+++ trunk/libjava/classpath/configure.ac	2006-07-02 19:18:28.913906000 +0200
@@ -445,7 +445,8 @@
       AC_CHECK_FILE([$EXTRA_QT_INCLUDE_DIR/QWidget],
 	QT_CFLAGS="$QT_CFLAGS -I$EXTRA_QT_INCLUDE_DIR",
 	AC_MSG_WARN([QWidget not found])))
-	AC_CHECK_PROG(MOC, [moc], [moc])
+      QT4DIR=`$PKG_CONFIG --variable=prefix QtGui`
+      AC_CHECK_PROG(MOC, [moc], [$QT4DIR/bin/moc], [], $QT4DIR/bin)
     fi
     if test "x$HAVE_QT4" = "xno"; then
       AC_MSG_NOTICE([Looking for QT_CFLAGS and QT_LIBS without pkg-config])
--- trunk/libjava/classpath/native/jni/qt-peer/Makefile.am.orig	2006-07-02 15:31:56.000000000 +0200
+++ trunk/libjava/classpath/native/jni/qt-peer/Makefile.am	2006-07-02 19:55:57.446430500 +0200
@@ -8,7 +8,10 @@
 gcjversionedlib_LTLIBRARIES = libqtpeer.la
 
 AM_LDFLAGS = @CLASSPATH_MODULE@ @QT_LIBS@
-AM_CPPFLAGS = @CLASSPATH_INCLUDES@
+AM_CPPFLAGS = @CLASSPATH_INCLUDES@ \
+	-I$(top_builddir)/../../libstdc++-v3/include \
+	-I$(top_builddir)/../../libstdc++-v3/include/$(target_alias) \
+	-I$(top_srcdir)/../../libstdc++-v3/libsupc++
 
 AM_CXXFLAGS = @QT_CFLAGS@
 
--- trunk/libjava/classpath/native/jni/qt-peer/Makefile.in.orig	2006-07-02 15:31:56.000000000 +0200
+++ trunk/libjava/classpath/native/jni/qt-peer/Makefile.in	2006-07-02 19:56:14.711509500 +0200
@@ -324,7 +324,10 @@
 gcjversionedlibdir = $(libdir)/gcj-$(gcc_version)
 gcjversionedlib_LTLIBRARIES = libqtpeer.la
 AM_LDFLAGS = @CLASSPATH_MODULE@ @QT_LIBS@
-AM_CPPFLAGS = @CLASSPATH_INCLUDES@
+AM_CPPFLAGS = @CLASSPATH_INCLUDES@ \
+	-I$(top_builddir)/../../libstdc++-v3/include \
+	-I$(top_builddir)/../../libstdc++-v3/include/$(target_alias) \
+	-I$(top_srcdir)/../../libstdc++-v3/libsupc++
 AM_CXXFLAGS = @QT_CFLAGS@
 libqtpeer_la_MOC = \
 	slotcallbacks.moc.h
--- trunk/libjava/classpath/native/jni/qt-peer/componentevent.cpp.orig	2005-11-16 00:20:01.000000000 +0100
+++ trunk/libjava/classpath/native/jni/qt-peer/componentevent.cpp	2006-02-26 17:59:49.285101500 +0100
@@ -39,6 +39,7 @@
 #include <QWidget>
 #include <QPoint>
 
+#include <assert.h>
 #include "componentevent.h"
   
 AWTInitEvent::AWTInitEvent(JNIEnv *env, jobject obj) : AWTEvent()
--- trunk/libjava/classpath/gnu/java/awt/peer/qt/QtToolkit.java.orig	2006-07-02 15:31:57.000000000 +0200
+++ trunk/libjava/classpath/gnu/java/awt/peer/qt/QtToolkit.java	2006-07-02 18:52:59.444756500 +0200
@@ -136,7 +136,7 @@
   {
     eventQueue = new EventQueue();
     repaintThread = new QtRepaintThread();
-    System.loadLibrary("qtpeer");
+    System.loadLibrary("-gnu-java-awt-peer-qt");
 
     String theme = null;
     try 
