--- gcc-3.3/gcc/ada/Make-lang.in~	Fri May 16 15:14:53 2003
+++ gcc-3.3/gcc/ada/Make-lang.in	Fri May 16 15:18:07 2003
@@ -347,7 +347,7 @@
 ada/doctools/xgnatug : ada/xgnatug.adb
 	-$(MKDIR) ada/doctools
 	cp $^ ada/doctools
-	cd ada/doctools && gnatmake -q xgnatug
+	cd ada/doctools && gnatmake -q xgnatug -aO../rts
 
 $(srcdir)/ada/gnat_ug_unx.texi : ada/doctools/xgnatug \
 	$(srcdir)/ada/gnat_ug.texi $(srcdir)/ada/ug_words
@@ -871,27 +871,27 @@
 ada/treeprs.ads : ada/treeprs.adt ada/sinfo.ads ada/xtreeprs.adb
 	-$(MKDIR) ada/bldtools
 	$(CP) $^ ada/bldtools
-	(cd ada/bldtools; gnatmake -q xtreeprs ; ./xtreeprs ../treeprs.ads )
+	(cd ada/bldtools; gnatmake -q xtreeprs -aO../rts ; ./xtreeprs ../treeprs.ads )
 
 ada/einfo.h : ada/einfo.ads ada/einfo.adb ada/xeinfo.adb
 	-$(MKDIR) ada/bldtools
 	$(CP) $^ ada/bldtools
-	(cd ada/bldtools; gnatmake -q xeinfo ; ./xeinfo ../einfo.h )
+	(cd ada/bldtools; gnatmake -q xeinfo -aO../rts ; ./xeinfo ../einfo.h )
 
 ada/sinfo.h : ada/sinfo.ads ada/xsinfo.adb
 	-$(MKDIR) ada/bldtools
 	$(CP) $^ ada/bldtools
-	(cd ada/bldtools; gnatmake -q xsinfo ; ./xsinfo ../sinfo.h )
+	(cd ada/bldtools; gnatmake -q xsinfo -aO../rts ; ./xsinfo ../sinfo.h )
 
 ada/nmake.adb : ada/sinfo.ads ada/nmake.adt ada/xnmake.adb
 	-$(MKDIR) ada/bldtools
 	$(CP) $^ ada/bldtools
-	(cd ada/bldtools; gnatmake -q xnmake ; ./xnmake -b ../nmake.adb )
+	(cd ada/bldtools; gnatmake -q xnmake -aO../rts ; ./xnmake -b ../nmake.adb )
 
 ada/nmake.ads :  ada/sinfo.ads ada/nmake.adt ada/xnmake.adb ada/nmake.adb
 	-$(MKDIR) ada/bldtools
 	$(CP) $^ ada/bldtools
-	(cd ada/bldtools; gnatmake -q xnmake ; ./xnmake -s ../nmake.ads )
+	(cd ada/bldtools; gnatmake -q xnmake -aO../rts ; ./xnmake -s ../nmake.ads )
 
 update-sources : ada/treeprs.ads ada/einfo.h ada/sinfo.h ada/nmake.adb \
 	ada/nmake.ads
