--- trunk/gcc/config/rs6000/rs6000.c    (revision 115196)
+++ trunk/gcc/config/rs6000/rs6000.c    (working copy)
@@ -3522,6 +3522,7 @@
   if ((GET_CODE (x) == PRE_INC || GET_CODE (x) == PRE_DEC)
       && !ALTIVEC_VECTOR_MODE (mode)
       && !SPE_VECTOR_MODE (mode)
+      && mode != TFmode
       /* Restrict addressing for DI because of our SUBREG hackery.  */
       && !(TARGET_E500_DOUBLE && (mode == DFmode || mode == DImode))
       && TARGET_UPDATE
