--- gcc-4.7.1/config/cloog.m4.orig	2011-03-02 04:48:36.000000000 +0100
+++ gcc-4.7.1/config/cloog.m4	2012-07-06 18:44:25.045888239 +0200
@@ -207,8 +207,8 @@
 m4_define([_CLOOG_CHECK_CT_PROG],[AC_LANG_PROGRAM(
   [#include "cloog/cloog.h"],
   [#if CLOOG_VERSION_MAJOR != $1 \
-    || CLOOG_VERSION_MINOR != $2 \
-    || CLOOG_VERSION_REVISION < $3
+    || CLOOG_VERSION_MINOR < $2 \
+    || (CLOOG_VERSION_MINOR == $2 && CLOOG_VERSION_REVISION < $3)
     choke me
    #endif])])
 
--- gcc-4.7.1/configure.orig	2012-02-02 11:20:32.000000000 +0100
+++ gcc-4.7.1/configure	2012-07-06 19:07:17.112497327 +0200
@@ -5973,8 +5973,8 @@
 main ()
 {
 #if CLOOG_VERSION_MAJOR != 0 \
-    || CLOOG_VERSION_MINOR != 17 \
-    || CLOOG_VERSION_REVISION < 0
+    || CLOOG_VERSION_MINOR < 17 \
+    || (CLOOG_VERSION_MINOR == 17 && CLOOG_VERSION_REVISION < 0)
     choke me
    #endif
   ;
