--- gcc-3.2/libstdc++-v3/configure.orig	Sat Jul 27 01:23:04 2002
+++ gcc-3.2/libstdc++-v3/configure	Thu Aug 22 08:03:18 2002
@@ -2957,7 +2957,7 @@
               int j;
               __locale_t        loc;
                __locale_t        loc_dup;
-              loc = __newlocale(1 << LC_ALL, "de_DE", 0);
+              loc = __newlocale(1 << LC_ALL, "C", 0);
               loc_dup = __duplocale(loc);
               i = __strcoll_l(__one, __two, loc);
               j = __strcoll_l(__one, __two, loc_dup);
