--- gcc/libjava/configure.ac	2006-01-23 18:53:45.000000000 +0100
+++ gcc/libjava/configure.ac	2006-02-26 16:02:13.828163000 +0100
@@ -84,6 +84,21 @@
   *)   AC_MSG_ERROR(bad value ${enableval} for multilib option) ;;
  esac], [multilib=yes])dnl
 
+# Default to --enable-libjava-multilib
+AC_ARG_ENABLE(libjava-multilib,
+  AS_HELP_STRING([--enable-libjava-multilib],
+                 [build many library versions (default)]),
+[case "${enableval}" in
+  yes) multilib=yes ;;
+  no)  multilib=no ;;
+  *)   AC_MSG_ERROR(bad value ${enableval} for libjava-multilib option) ;;
+ esac], [multilib=yes])dnl
+if test "$multilib" = no; then
+  # Reset also --enable-multilib state, as that is what is looked at
+  # by config-ml.in
+  ac_configure_args="$ac_configure_args --disable-multilib"
+fi
+
 # It may not be safe to run linking tests in AC_PROG_CC/AC_PROG_CXX.
 GCC_NO_EXECUTABLES
 
--- gcc-4_2-branch/libjava/classpath/ltconfig.orig	2006-10-21 14:45:42.000000000 +0000
+++ gcc-4_2-branch/libjava/classpath/ltconfig	2006-12-29 18:28:13.000000000 +0000
@@ -1272,6 +1272,27 @@
   # people can always --disable-shared, the test was removed, and we
   # assume the GNU/Linux dynamic linker is in use.
   dynamic_linker='GNU/Linux ld.so'
+
+
+  # Find out which ABI we are using (multilib Linux x86_64 hack).
+  libsuff=
+  case "$host_cpu" in
+  x86_64*|s390x*|powerpc64*|sparc64*)
+    echo '#line 1281 "ltconfig"' > conftest.$ac_ext
+    if (eval $ac_compile) && test -s conftest.$objext; then
+      case `/usr/bin/file conftest.$objext` in
+      *64-bit*)
+        libsuff=64
+        ;;
+      esac
+    fi
+    rm -rf conftest*
+    ;;
+  *)
+    ;;
+  esac
+  sys_lib_dlsearch_path_spec="/lib${libsuff} /usr/lib${libsuff}"
+  sys_lib_search_path_spec="/lib${libsuff} /usr/lib${libsuff} /usr/local/lib${libsuff}"
   ;;
 
 netbsd*)
