--- gcc/gcc/ipa-reference.c	(revision 107400)
+++ gcc/gcc/ipa-reference.c	(working copy)
@@ -550,6 +550,8 @@ scan_for_static_refs (tree *tp, 
     case ADDR_EXPR:
       /* This case is here to find addresses on rhs of constructors in
 	 decl_initial of static variables. */
+      if (TREE_CODE (TREE_OPERAND (t, 0)) == CONSTRUCTOR)
+	break;
       check_rhs_var (local, t);
       *walk_subtrees = 0;
       break;
--- gcc/gcc/ipa-type-escape.c	(revision 107400)
+++ gcc/gcc/ipa-type-escape.c	(working copy)
@@ -1284,6 +1284,8 @@ scan_for_refs (tree *tp, int *walk_subtr
     case ADDR_EXPR:
       /* This case is here to find addresses on rhs of constructors in
 	 decl_initial of static variables. */
+      if (TREE_CODE (TREE_OPERAND (t, 0)) == CONSTRUCTOR)
+	break;
       check_rhs_var (t);
       *walk_subtrees = 0;
       break;
