--- gcc/libstdc++-v3/include/ext/algorithm.vis1~	2005-10-31 21:15:57.000000000 +0100
+++ gcc/libstdc++-v3/include/ext/algorithm	2005-11-03 10:28:32.000000000 +0100
@@ -1,6 +1,6 @@
 // Algorithm extensions -*- C++ -*-
 
-// Copyright (C) 2001, 2002, 2004 Free Software Foundation, Inc.
+// Copyright (C) 2001, 2002, 2004, 2005 Free Software Foundation, Inc.
 //
 // This file is part of the GNU ISO C++ Library.  This library is free
 // software; you can redistribute it and/or modify it under the
@@ -62,6 +62,7 @@
 #define _EXT_ALGORITHM 1
 
 #pragma GCC system_header
+#pragma GCC visibility push(default)
 
 #include <algorithm>
 
@@ -523,4 +524,6 @@
     }
 } // namespace __gnu_cxx
 
+#pragma GCC visibility pop
+
 #endif /* _EXT_ALGORITHM */
--- gcc/libstdc++-v3/include/ext/array_allocator.h.vis1~	2005-10-31 21:15:57.000000000 +0100
+++ gcc/libstdc++-v3/include/ext/array_allocator.h	2005-11-03 10:28:32.000000000 +0100
@@ -34,6 +34,8 @@
 #ifndef _ARRAY_ALLOCATOR_H
 #define _ARRAY_ALLOCATOR_H 1
 
+#pragma GCC visibility push(default)
+
 #include <cstddef>
 #include <new>
 #include <bits/functexcept.h>
@@ -142,4 +144,6 @@
     { return false; }
 } // namespace __gnu_cxx
 
+#pragma GCC visibility pop
+
 #endif
--- gcc/libstdc++-v3/include/ext/bitmap_allocator.h.vis1~	2005-10-31 21:15:57.000000000 +0100
+++ gcc/libstdc++-v3/include/ext/bitmap_allocator.h	2005-11-03 10:28:32.000000000 +0100
@@ -34,6 +34,8 @@
 #ifndef _BITMAP_ALLOCATOR_H
 #define _BITMAP_ALLOCATOR_H 1
 
+#pragma GCC visibility push(default)
+
 // For std::size_t, and ptrdiff_t.
 #include <cstddef>
 
@@ -1289,6 +1291,8 @@
 
 }
 
+#pragma GCC visibility pop
+
 #endif 
 
 //  LocalWords:  namespace GTHREADS bool const gthread endif Mutex mutex
--- gcc/libstdc++-v3/include/ext/debug_allocator.h.vis1~	2005-10-31 21:15:57.000000000 +0100
+++ gcc/libstdc++-v3/include/ext/debug_allocator.h	2005-11-03 10:28:32.000000000 +0100
@@ -48,6 +48,8 @@
 #ifndef _DEBUG_ALLOCATOR_H
 #define _DEBUG_ALLOCATOR_H 1
 
+#pragma GCC visibility push(default)
+
 #include <stdexcept>
 
 namespace __gnu_cxx
@@ -122,4 +124,6 @@
     };
 } // namespace __gnu_cxx
 
+#pragma GCC visibility pop
+
 #endif
--- gcc/libstdc++-v3/include/ext/functional.vis1~	2005-10-31 21:15:57.000000000 +0100
+++ gcc/libstdc++-v3/include/ext/functional	2005-11-03 10:28:32.000000000 +0100
@@ -1,6 +1,6 @@
 // Functional extensions -*- C++ -*-
 
-// Copyright (C) 2002, 2004 Free Software Foundation, Inc.
+// Copyright (C) 2002, 2004, 2005 Free Software Foundation, Inc.
 //
 // This file is part of the GNU ISO C++ Library.  This library is free
 // software; you can redistribute it and/or modify it under the
@@ -62,6 +62,7 @@
 #define _EXT_FUNCTIONAL 1
 
 #pragma GCC system_header
+#pragma GCC visibility push(default)
 
 #include <functional>
 
@@ -421,5 +422,8 @@
     mem_fun1_ref(_Ret (_Tp::*__f)(_Arg) const)
     { return const_mem_fun1_ref_t<_Ret, _Tp, _Arg>(__f); }
 } // namespace __gnu_cxx
+
+#pragma GCC visibility pop
+
 #endif
 
--- gcc/libstdc++-v3/include/ext/hash_map.vis1~	2005-10-31 21:15:57.000000000 +0100
+++ gcc/libstdc++-v3/include/ext/hash_map	2005-11-03 10:30:29.000000000 +0100
@@ -61,6 +61,8 @@
 #ifndef _HASH_MAP
 #define _HASH_MAP 1
 
+#pragma GCC visibility push(default)
+
 #include <ext/hashtable.h>
 #include <bits/concept_check.h>
 
@@ -622,4 +624,6 @@
 # include <debug/hash_map>
 #endif
 
+#pragma GCC visibility pop
+
 #endif
--- gcc/libstdc++-v3/include/ext/hash_set.vis1~	2005-10-31 21:15:57.000000000 +0100
+++ gcc/libstdc++-v3/include/ext/hash_set	2005-11-03 10:30:57.000000000 +0100
@@ -61,6 +61,8 @@
 #ifndef _HASH_SET
 #define _HASH_SET 1
 
+#pragma GCC visibility push(default)
+
 #include <ext/hashtable.h>
 #include <bits/concept_check.h>
 
@@ -588,4 +590,6 @@
 # include <debug/hash_set>
 #endif
 
+#pragma GCC visibility pop
+
 #endif
--- gcc/libstdc++-v3/include/ext/iterator.vis1~	2005-10-31 21:15:57.000000000 +0100
+++ gcc/libstdc++-v3/include/ext/iterator	2005-11-03 10:28:34.000000000 +0100
@@ -1,6 +1,6 @@
 // HP/SGI iterator extensions -*- C++ -*-
 
-// Copyright (C) 2001, 2002, 2004 Free Software Foundation, Inc.
+// Copyright (C) 2001, 2002, 2004, 2005 Free Software Foundation, Inc.
 //
 // This file is part of the GNU ISO C++ Library.  This library is free
 // software; you can redistribute it and/or modify it under the
@@ -62,6 +62,7 @@
 #define _EXT_ITERATOR 1
 
 #pragma GCC system_header
+#pragma GCC visibility push(default)
 
 #include <bits/concept_check.h>
 #include <iterator>
@@ -113,5 +114,7 @@
     }
 } // namespace __gnu_cxx
 
+#pragma GCC visibility pop
+
 #endif
 
--- gcc/libstdc++-v3/include/ext/malloc_allocator.h.vis1~	2005-10-31 21:15:57.000000000 +0100
+++ gcc/libstdc++-v3/include/ext/malloc_allocator.h	2005-11-03 10:28:34.000000000 +0100
@@ -1,6 +1,7 @@
 // Allocator that wraps "C" malloc -*- C++ -*-
 
-// Copyright (C) 2001, 2002, 2003, 2004 Free Software Foundation, Inc.
+// Copyright (C) 2001, 2002, 2003, 2004, 2005
+// Free Software Foundation, Inc.
 //
 // This file is part of the GNU ISO C++ Library.  This library is free
 // software; you can redistribute it and/or modify it under the
@@ -34,6 +35,8 @@
 #ifndef _MALLOC_ALLOCATOR_H
 #define _MALLOC_ALLOCATOR_H 1
 
+#pragma GCC visibility push(default)
+
 #include <cstdlib>
 #include <new>
 #include <bits/functexcept.h>
@@ -122,4 +125,6 @@
     { return false; }
 } // namespace __gnu_cxx
 
+#pragma GCC visibility pop
+
 #endif
--- gcc/libstdc++-v3/include/ext/memory.vis1~	2005-10-31 21:15:57.000000000 +0100
+++ gcc/libstdc++-v3/include/ext/memory	2005-11-03 10:28:34.000000000 +0100
@@ -1,6 +1,6 @@
 // Memory extensions -*- C++ -*-
 
-// Copyright (C) 2002, 2004 Free Software Foundation, Inc.
+// Copyright (C) 2002, 2004, 2005 Free Software Foundation, Inc.
 //
 // This file is part of the GNU ISO C++ Library.  This library is free
 // software; you can redistribute it and/or modify it under the
@@ -62,6 +62,7 @@
 #define _EXT_MEMORY 1
 
 #pragma GCC system_header
+#pragma GCC visibility push(default)
 
 #include <memory>
 #include <bits/stl_tempbuf.h>
@@ -194,5 +195,7 @@
     };
 } // namespace __gnu_cxx
 
+#pragma GCC visibility pop
+
 #endif
 
--- gcc/libstdc++-v3/include/ext/mt_allocator.h.vis1~	2005-10-31 21:15:57.000000000 +0100
+++ gcc/libstdc++-v3/include/ext/mt_allocator.h	2005-11-03 10:28:34.000000000 +0100
@@ -34,6 +34,8 @@
 #ifndef _MT_ALLOCATOR_H
 #define _MT_ALLOCATOR_H 1
 
+#pragma GCC visibility push(default)
+
 #include <new>
 #include <cstdlib>
 #include <bits/functexcept.h>
@@ -731,4 +733,6 @@
 #undef __thread_default
 } // namespace __gnu_cxx
 
+#pragma GCC visibility pop
+
 #endif
--- gcc/libstdc++-v3/include/ext/new_allocator.h.vis1~	2005-10-31 21:15:57.000000000 +0100
+++ gcc/libstdc++-v3/include/ext/new_allocator.h	2005-11-03 10:28:34.000000000 +0100
@@ -1,6 +1,6 @@
 // Allocator that wraps operator new -*- C++ -*-
 
-// Copyright (C) 2001, 2002, 2003, 2004 Free Software Foundation, Inc.
+// Copyright (C) 2001, 2002, 2003, 2004, 2005 Free Software Foundation, Inc.
 //
 // This file is part of the GNU ISO C++ Library.  This library is free
 // software; you can redistribute it and/or modify it under the
@@ -34,6 +34,8 @@
 #ifndef _NEW_ALLOCATOR_H
 #define _NEW_ALLOCATOR_H 1
 
+#pragma GCC visibility push(default)
+
 #include <new>
 #include <bits/functexcept.h>
 
@@ -118,4 +120,6 @@
     { return false; }
 } // namespace __gnu_cxx
 
+#pragma GCC visibility pop
+
 #endif
--- gcc/libstdc++-v3/include/ext/numeric.vis1~	2005-10-31 21:15:57.000000000 +0100
+++ gcc/libstdc++-v3/include/ext/numeric	2005-11-03 10:28:34.000000000 +0100
@@ -1,6 +1,6 @@
 // Numeric extensions -*- C++ -*-
 
-// Copyright (C) 2002, 2004 Free Software Foundation, Inc.
+// Copyright (C) 2002, 2004, 2005 Free Software Foundation, Inc.
 //
 // This file is part of the GNU ISO C++ Library.  This library is free
 // software; you can redistribute it and/or modify it under the
@@ -62,6 +62,7 @@
 #define _EXT_NUMERIC 1
 
 #pragma GCC system_header
+#pragma GCC visibility push(default)
 
 #include <bits/concept_check.h>
 #include <numeric>
@@ -146,5 +147,7 @@
     }
 } // namespace __gnu_cxx
 
+#pragma GCC visibility pop
+
 #endif
 
--- gcc/libstdc++-v3/include/ext/pod_char_traits.h.vis1~	2005-10-31 21:15:57.000000000 +0100
+++ gcc/libstdc++-v3/include/ext/pod_char_traits.h	2005-11-03 10:28:34.000000000 +0100
@@ -37,6 +37,8 @@
 #ifndef _POD_CHAR_TRAITS_H
 #define _POD_CHAR_TRAITS_H 1
 
+#pragma GCC visibility push(default)
+
 #include <string>
 
 namespace __gnu_cxx
@@ -182,4 +184,6 @@
     };
 }
 
+#pragma GCC visibility pop
+
 #endif
--- gcc/libstdc++-v3/include/ext/pool_allocator.h.vis1~	2005-10-31 21:15:57.000000000 +0100
+++ gcc/libstdc++-v3/include/ext/pool_allocator.h	2005-11-03 10:28:34.000000000 +0100
@@ -1,6 +1,6 @@
 // Allocators -*- C++ -*-
 
-// Copyright (C) 2001, 2002, 2003, 2004 Free Software Foundation, Inc.
+// Copyright (C) 2001, 2002, 2003, 2004, 2005 Free Software Foundation, Inc.
 //
 // This file is part of the GNU ISO C++ Library.  This library is free
 // software; you can redistribute it and/or modify it under the
@@ -47,6 +47,8 @@
 #ifndef _POOL_ALLOCATOR_H
 #define _POOL_ALLOCATOR_H 1
 
+#pragma GCC visibility push(default)
+
 #include <bits/c++config.h>
 #include <cstdlib>
 #include <new>
@@ -252,4 +254,6 @@
     }
 } // namespace __gnu_cxx
 
+#pragma GCC visibility pop
+
 #endif
--- gcc/libstdc++-v3/include/ext/rb_tree.vis1~	2005-10-31 21:15:57.000000000 +0100
+++ gcc/libstdc++-v3/include/ext/rb_tree	2005-11-03 10:28:34.000000000 +0100
@@ -1,6 +1,6 @@
 // rb_tree extension -*- C++ -*-
 
-// Copyright (C) 2002, 2003, 2004 Free Software Foundation, Inc.
+// Copyright (C) 2002, 2003, 2004, 2005 Free Software Foundation, Inc.
 //
 // This file is part of the GNU ISO C++ Library.  This library is free
 // software; you can redistribute it and/or modify it under the
@@ -62,6 +62,7 @@
 #define _RB_TREE 1
 
 #pragma GCC system_header
+#pragma GCC visibility push(default)
 
 #include <bits/stl_tree.h>
 
@@ -94,4 +95,6 @@
     };
 } // namespace __gnu_cxx
 
+#pragma GCC visibility pop
+
 #endif
--- gcc/libstdc++-v3/include/ext/rope.vis1~	2005-10-31 21:15:57.000000000 +0100
+++ gcc/libstdc++-v3/include/ext/rope	2005-11-03 10:28:34.000000000 +0100
@@ -48,6 +48,8 @@
 #ifndef _ROPE
 #define _ROPE 1
 
+#pragma GCC visibility push(default)
+
 #include <bits/stl_algobase.h>
 #include <bits/stl_construct.h>
 #include <bits/stl_uninitialized.h>
@@ -2898,4 +2900,6 @@
 
 # include <ext/ropeimpl.h>
 
+#pragma GCC visibility pop
+
 #endif
--- gcc/libstdc++-v3/include/ext/slist.vis1~	2005-10-31 21:15:57.000000000 +0100
+++ gcc/libstdc++-v3/include/ext/slist	2005-11-03 10:28:34.000000000 +0100
@@ -49,6 +49,8 @@
 #ifndef _SLIST
 #define _SLIST 1
 
+#pragma GCC visibility push(default)
+
 #include <bits/stl_algobase.h>
 #include <bits/allocator.h>
 #include <bits/stl_construct.h>
@@ -1076,4 +1078,7 @@
 };
 
 } // namespace std
+
+#pragma GCC visibility pop
+
 #endif
--- gcc/libstdc++-v3/include/ext/stdio_filebuf.h.vis1~	2005-10-31 21:15:57.000000000 +0100
+++ gcc/libstdc++-v3/include/ext/stdio_filebuf.h	2005-11-03 10:28:34.000000000 +0100
@@ -1,6 +1,6 @@
 // File descriptor layer for filebuf -*- C++ -*-
 
-// Copyright (C) 2002, 2003, 2004 Free Software Foundation, Inc.
+// Copyright (C) 2002, 2003, 2004, 2005 Free Software Foundation, Inc.
 //
 // This file is part of the GNU ISO C++ Library.  This library is free
 // software; you can redistribute it and/or modify it under the
@@ -35,6 +35,7 @@
 #define _STDIO_FILEBUF_H 1
 
 #pragma GCC system_header
+#pragma GCC visibility push(default)
 
 #include <fstream>
 
@@ -158,4 +159,6 @@
     }
 } // namespace __gnu_cxx
 
+#pragma GCC visibility pop
+
 #endif
--- gcc/libstdc++-v3/include/ext/stdio_sync_filebuf.h.vis1~	2005-10-31 21:15:57.000000000 +0100
+++ gcc/libstdc++-v3/include/ext/stdio_sync_filebuf.h	2005-11-03 10:28:34.000000000 +0100
@@ -1,6 +1,6 @@
 // Iostreams wrapper for stdio FILE* -*- C++ -*-
 
-// Copyright (C) 2003, 2004 Free Software Foundation, Inc.
+// Copyright (C) 2003, 2004, 2005 Free Software Foundation, Inc.
 //
 // This file is part of the GNU ISO C++ Library.  This library is free
 // software; you can redistribute it and/or modify it under the
@@ -35,6 +35,7 @@
 #define _STDIO_SYNC_FILEBUF_H 1
 
 #pragma GCC system_header
+#pragma GCC visibility push(default)
 
 #include <streambuf>
 #include <unistd.h>
@@ -279,4 +280,6 @@
 #endif
 } // namespace __gnu_cxx
 
+#pragma GCC visibility pop
+
 #endif
--- gcc/libstdc++-v3/include/debug/bitset.vis1~	2005-10-31 21:16:05.000000000 +0100
+++ gcc/libstdc++-v3/include/debug/bitset	2005-11-03 10:27:12.000000000 +0100
@@ -31,6 +31,8 @@
 #ifndef _GLIBCXX_DEBUG_BITSET
 #define _GLIBCXX_DEBUG_BITSET
 
+#pragma GCC visibility push(default)
+
 #include <bitset>
 #include <debug/safe_sequence.h>
 #include <debug/safe_iterator.h>
@@ -318,4 +320,6 @@
     { return __os << __x._M_base(); }
 } // namespace __gnu_debug_def
 
+#pragma GCC visibility pop
+
 #endif
--- gcc/libstdc++-v3/include/debug/deque.vis1~	2005-10-31 21:16:05.000000000 +0100
+++ gcc/libstdc++-v3/include/debug/deque	2005-11-03 10:28:26.000000000 +0100
@@ -31,6 +31,8 @@
 #ifndef _GLIBCXX_DEBUG_DEQUE
 #define _GLIBCXX_DEBUG_DEQUE 1
 
+#pragma GCC visibility push(default)
+
 #include <deque>
 #include <debug/safe_sequence.h>
 #include <debug/safe_iterator.h>
@@ -383,4 +385,6 @@
     { __lhs.swap(__rhs); }
 } // namespace __gnu_debug_def
 
+#pragma GCC visibility pop
+
 #endif
--- gcc/libstdc++-v3/include/debug/hash_map.vis1~	2005-10-31 21:16:05.000000000 +0100
+++ gcc/libstdc++-v3/include/debug/hash_map	2005-11-03 10:31:55.000000000 +0100
@@ -31,8 +31,12 @@
 #ifndef _GLIBCXX_DEBUG_HASH_MAP
 #define _GLIBCXX_DEBUG_HASH_MAP 1
 
+#pragma GCC visibility push(default)
+
 #include <ext/hash_map>
 #include <debug/hash_map.h>
 #include <debug/hash_multimap.h>
 
+#pragma GCC visibility pop
+
 #endif
--- gcc/libstdc++-v3/include/debug/hash_set.vis1~	2005-10-31 21:16:05.000000000 +0100
+++ gcc/libstdc++-v3/include/debug/hash_set	2005-11-03 10:32:22.000000000 +0100
@@ -31,8 +31,12 @@
 #ifndef _GLIBCXX_DEBUG_HASH_SET
 #define _GLIBCXX_DEBUG_HASH_SET 1
 
+#pragma GCC visibility push(default)
+
 #include <ext/hash_set>
 #include <debug/hash_set.h>
 #include <debug/hash_multiset.h>
 
+#pragma GCC visibility pop
+
 #endif
--- gcc/libstdc++-v3/include/debug/list.vis1~	2005-10-31 21:16:05.000000000 +0100
+++ gcc/libstdc++-v3/include/debug/list	2005-11-03 10:28:28.000000000 +0100
@@ -31,6 +31,8 @@
 #ifndef _GLIBCXX_DEBUG_LIST
 #define _GLIBCXX_DEBUG_LIST 1
 
+#pragma GCC visibility push(default)
+
 #include <list>
 #include <bits/stl_algo.h>
 #include <debug/safe_sequence.h>
@@ -502,4 +504,6 @@
     { __lhs.swap(__rhs); }
 } // namespace __gnu_debug_def
 
+#pragma GCC visibility pop
+
 #endif
--- gcc/libstdc++-v3/include/debug/map.vis1~	2005-10-31 21:16:05.000000000 +0100
+++ gcc/libstdc++-v3/include/debug/map	2005-11-03 10:28:28.000000000 +0100
@@ -1,6 +1,6 @@
 // Debugging map/multimap implementation -*- C++ -*-
 
-// Copyright (C) 2003
+// Copyright (C) 2003, 2005
 // Free Software Foundation, Inc.
 //
 // This file is part of the GNU ISO C++ Library.  This library is free
@@ -31,8 +31,12 @@
 #ifndef _GLIBCXX_DEBUG_MAP
 #define _GLIBCXX_DEBUG_MAP 1
 
+#pragma GCC visibility push(default)
+
 #include <map>
 #include <debug/map.h>
 #include <debug/multimap.h>
 
+#pragma GCC visibility pop
+
 #endif
--- gcc/libstdc++-v3/include/debug/set.vis1~	2005-10-31 21:16:05.000000000 +0100
+++ gcc/libstdc++-v3/include/debug/set	2005-11-03 10:28:28.000000000 +0100
@@ -1,6 +1,6 @@
 // Debugging set/multiset implementation -*- C++ -*-
 
-// Copyright (C) 2003
+// Copyright (C) 2003, 2005
 // Free Software Foundation, Inc.
 //
 // This file is part of the GNU ISO C++ Library.  This library is free
@@ -31,8 +31,12 @@
 #ifndef _GLIBCXX_DEBUG_SET
 #define _GLIBCXX_DEBUG_SET 1
 
+#pragma GCC visibility push(default)
+
 #include <set>
 #include <debug/set.h>
 #include <debug/multiset.h>
 
+#pragma GCC visibility pop
+
 #endif
--- gcc/libstdc++-v3/include/debug/string.vis1~	2005-10-31 21:16:05.000000000 +0100
+++ gcc/libstdc++-v3/include/debug/string	2005-11-03 10:28:30.000000000 +0100
@@ -31,6 +31,8 @@
 #ifndef _GLIBCXX_DEBUG_STRING
 #define _GLIBCXX_DEBUG_STRING 1
 
+#pragma GCC visibility push(default)
+
 #include <string>
 #include <debug/safe_sequence.h>
 #include <debug/safe_iterator.h>
@@ -1015,4 +1017,6 @@
 
 } // namespace __gnu_debug
 
+#pragma GCC visibility pop
+
 #endif
--- gcc/libstdc++-v3/include/debug/vector.vis1~	2005-10-31 21:16:05.000000000 +0100
+++ gcc/libstdc++-v3/include/debug/vector	2005-11-03 10:28:32.000000000 +0100
@@ -31,6 +31,8 @@
 #ifndef _GLIBCXX_DEBUG_VECTOR
 #define _GLIBCXX_DEBUG_VECTOR 1
 
+#pragma GCC visibility push(default)
+
 #include <vector>
 #include <utility>
 #include <debug/safe_sequence.h>
@@ -413,4 +415,6 @@
     { __lhs.swap(__rhs); }
 } // namespace __gnu_debug_def
 
+#pragma GCC visibility pop
+
 #endif
--- gcc/libstdc++-v3/include/tr1/array.vis1~	2005-10-31 21:16:08.000000000 +0100
+++ gcc/libstdc++-v3/include/tr1/array	2005-11-03 10:28:38.000000000 +0100
@@ -34,6 +34,8 @@
 #ifndef _ARRAY
 #define _ARRAY 1
 
+#pragma GCC visibility push(default)
+
 #include <new>
 #include <iterator>
 #include <algorithm>
@@ -231,4 +233,6 @@
 } // namespace std::tr1
 }
 
+#pragma GCC visibility pop
+
 #endif
--- gcc/libstdc++-v3/include/tr1/functional.vis1~	2005-10-31 21:16:08.000000000 +0100
+++ gcc/libstdc++-v3/include/tr1/functional	2005-11-03 10:28:38.000000000 +0100
@@ -36,6 +36,8 @@
 
 #pragma GCC system_header
 
+#pragma GCC visibility push(default)
+
 #include "../functional"
 #include <typeinfo>
 #include <tr1/type_traits>
@@ -1270,4 +1272,6 @@
 }
 }
 
+#pragma GCC visibility pop
+
 #endif
--- gcc/libstdc++-v3/include/tr1/memory.vis1~	2005-10-31 21:16:08.000000000 +0100
+++ gcc/libstdc++-v3/include/tr1/memory	2005-11-03 10:28:38.000000000 +0100
@@ -30,6 +30,8 @@
 #ifndef _TR1_MEMORY
 #define _TR1_MEMORY 1
 
+#pragma GCC visibility push(default)
+
 #include "../memory"
 #include <functional>       // std::less
 #include <exception>        // std::exception
@@ -48,4 +50,6 @@
 
 #include <tr1/boost_shared_ptr.h>
 
+#pragma GCC visibility pop
+
 #endif
--- gcc/libstdc++-v3/include/tr1/tuple.vis1~	2005-10-31 21:16:08.000000000 +0100
+++ gcc/libstdc++-v3/include/tr1/tuple	2005-11-03 10:28:38.000000000 +0100
@@ -36,6 +36,8 @@
 #ifndef _TUPLE
 #define _TUPLE 1
 
+#pragma GCC visibility push(default)
+
 #include <tr1/utility>
 #include <tr1/ref_fwd.h>
 
@@ -272,4 +274,6 @@
 
 #include <tr1/functional>
 
+#pragma GCC visibility pop
+
 #endif
--- gcc/libstdc++-v3/include/tr1/type_traits.vis1~	2005-10-31 21:16:08.000000000 +0100
+++ gcc/libstdc++-v3/include/tr1/type_traits	2005-11-03 10:28:38.000000000 +0100
@@ -34,6 +34,8 @@
 #ifndef _TYPE_TRAITS
 #define _TYPE_TRAITS 1
 
+#pragma GCC visibility push(default)
+
 #include <bits/c++config.h>
 #include <tr1/type_traits_fwd.h>
 
@@ -703,4 +705,6 @@
 }
 }
 
+#pragma GCC visibility pop
+
 #endif
--- gcc/libstdc++-v3/include/tr1/unordered_map.vis1~	2005-10-31 21:16:08.000000000 +0100
+++ gcc/libstdc++-v3/include/tr1/unordered_map	2005-11-03 10:28:38.000000000 +0100
@@ -34,6 +34,8 @@
 #ifndef GNU_LIBSTDCXX_TR1_UNORDERED_MAP_
 #define GNU_LIBSTDCXX_TR1_UNORDERED_MAP_
 
+#pragma GCC visibility push(default)
+
 #include <tr1/hashtable>
 #include <tr1/functional>
 #include <tr1/functional>
@@ -167,4 +169,6 @@
 }
 }
 
+#pragma GCC visibility pop
+
 #endif /* GNU_LIBSTDCXX_TR1_UNORDERED_MAP_ */
--- gcc/libstdc++-v3/include/tr1/unordered_set.vis1~	2005-10-31 21:16:08.000000000 +0100
+++ gcc/libstdc++-v3/include/tr1/unordered_set	2005-11-03 10:28:38.000000000 +0100
@@ -34,6 +34,8 @@
 #ifndef GNU_LIBSTDCXX_TR1_UNORDERED_SET_
 #define GNU_LIBSTDCXX_TR1_UNORDERED_SET_
 
+#pragma GCC visibility push(default)
+
 #include <tr1/hashtable>
 #include <tr1/functional>
 #include <memory>
@@ -162,4 +164,6 @@
 }
 }
 
+#pragma GCC visibility pop
+
 #endif /* GNU_LIBSTDCXX_TR1_UNORDERED_SET_ */
--- gcc/libstdc++-v3/include/tr1/utility.vis1~	2005-10-31 21:16:08.000000000 +0100
+++ gcc/libstdc++-v3/include/tr1/utility	2005-11-03 10:28:38.000000000 +0100
@@ -34,6 +34,8 @@
 #ifndef _TR1_UTILITY
 #define _TR1_UTILITY 1
 
+#pragma GCC visibility push(default)
+
 #include "../utility"
 
 namespace std
@@ -93,6 +95,8 @@
      get(const pair<_Tp1, _Tp2>& __in)
      { return __pair_get<_Int>::__const_get(__in); }
 }
-} 
+}
+
+#pragma GCC visibility pop
 
 #endif
--- gcc/libstdc++-v3/include/std/std_algorithm.h.vis1~	2005-10-31 21:16:10.000000000 +0100
+++ gcc/libstdc++-v3/include/std/std_algorithm.h	2005-11-03 10:28:34.000000000 +0100
@@ -1,6 +1,6 @@
 // <algorithm> -*- C++ -*-
 
-// Copyright (C) 2001, 2002 Free Software Foundation, Inc.
+// Copyright (C) 2001, 2002, 2005 Free Software Foundation, Inc.
 //
 // This file is part of the GNU ISO C++ Library.  This library is free
 // software; you can redistribute it and/or modify it under the
@@ -61,10 +61,13 @@
 #define _GLIBCXX_ALGORITHM 1
 
 #pragma GCC system_header
+#pragma GCC visibility push(default)
 
 #include <bits/stl_algobase.h>
 #include <bits/stl_construct.h>
 #include <bits/stl_uninitialized.h>
 #include <bits/stl_algo.h>
 
+#pragma GCC visibility pop
+
 #endif /* _GLIBCXX_ALGORITHM */
--- gcc/libstdc++-v3/include/std/std_bitset.h.vis1~	2005-10-31 21:16:10.000000000 +0100
+++ gcc/libstdc++-v3/include/std/std_bitset.h	2005-11-03 10:28:34.000000000 +0100
@@ -48,6 +48,7 @@
 #define _GLIBCXX_BITSET 1
 
 #pragma GCC system_header
+#pragma GCC visibility push(default)
 
 #include <cstddef>     // For size_t
 #include <cstring>     // For memset
@@ -1300,4 +1301,6 @@
 # include <debug/bitset>
 #endif
 
+#pragma GCC visibility pop
+
 #endif /* _GLIBCXX_BITSET */
--- gcc/libstdc++-v3/include/std/std_complex.h.vis1~	2005-10-31 21:16:10.000000000 +0100
+++ gcc/libstdc++-v3/include/std/std_complex.h	2005-11-03 10:28:34.000000000 +0100
@@ -43,6 +43,7 @@
 #define _GLIBCXX_COMPLEX 1
 
 #pragma GCC system_header
+#pragma GCC visibility push(default)
 
 #include <bits/c++config.h>
 #include <bits/cpp_type_traits.h>
@@ -1485,4 +1486,6 @@
   : _M_value(__z.__rep()) { }
 } // namespace std
 
+#pragma GCC visibility pop
+
 #endif	/* _GLIBCXX_COMPLEX */
--- gcc/libstdc++-v3/include/std/std_deque.h.vis1~	2005-10-31 21:16:10.000000000 +0100
+++ gcc/libstdc++-v3/include/std/std_deque.h	2005-11-03 10:28:34.000000000 +0100
@@ -1,6 +1,6 @@
 // <deque> -*- C++ -*-
 
-// Copyright (C) 2001, 2002, 2003 Free Software Foundation, Inc.
+// Copyright (C) 2001, 2002, 2003, 2005 Free Software Foundation, Inc.
 //
 // This file is part of the GNU ISO C++ Library.  This library is free
 // software; you can redistribute it and/or modify it under the
@@ -61,6 +61,7 @@
 #define _GLIBCXX_DEQUE 1
 
 #pragma GCC system_header
+#pragma GCC visibility push(default)
 
 #include <bits/functexcept.h>
 #include <bits/stl_algobase.h>
@@ -77,4 +78,6 @@
 # include <debug/deque>
 #endif
 
+#pragma GCC visibility pop
+
 #endif /* _GLIBCXX_DEQUE */
--- gcc/libstdc++-v3/include/std/std_fstream.h.vis1~	2005-10-31 21:16:10.000000000 +0100
+++ gcc/libstdc++-v3/include/std/std_fstream.h	2005-11-03 10:28:34.000000000 +0100
@@ -40,6 +40,7 @@
 #define _GLIBCXX_FSTREAM 1
 
 #pragma GCC system_header
+#pragma GCC visibility push(default)
 
 #include <istream>
 #include <ostream>
@@ -782,4 +783,6 @@
 # include <bits/fstream.tcc>
 #endif
 
+#pragma GCC visibility pop
+
 #endif /* _GLIBCXX_FSTREAM */
--- gcc/libstdc++-v3/include/std/std_functional.h.vis1~	2005-10-31 21:16:10.000000000 +0100
+++ gcc/libstdc++-v3/include/std/std_functional.h	2005-11-03 10:28:34.000000000 +0100
@@ -1,6 +1,6 @@
 // <functional> -*- C++ -*-
 
-// Copyright (C) 2001, 2002 Free Software Foundation, Inc.
+// Copyright (C) 2001, 2002, 2005 Free Software Foundation, Inc.
 //
 // This file is part of the GNU ISO C++ Library.  This library is free
 // software; you can redistribute it and/or modify it under the
@@ -49,9 +49,12 @@
 #define _GLIBCXX_FUNCTIONAL 1
 
 #pragma GCC system_header
+#pragma GCC visibility push(default)
 
 #include <bits/c++config.h>
 #include <cstddef>
 #include <bits/stl_function.h>
 
+#pragma GCC visibility pop
+
 #endif /* _GLIBCXX_FUNCTIONAL */
--- gcc/libstdc++-v3/include/std/std_iomanip.h.vis1~	2005-10-31 21:16:10.000000000 +0100
+++ gcc/libstdc++-v3/include/std/std_iomanip.h	2005-11-03 10:28:34.000000000 +0100
@@ -1,6 +1,6 @@
 // Standard stream manipulators -*- C++ -*-
 
-// Copyright (C) 1997, 1998, 1999, 2001, 2002, 2003
+// Copyright (C) 1997, 1998, 1999, 2001, 2002, 2003, 2005
 // Free Software Foundation, Inc.
 //
 // This file is part of the GNU ISO C++ Library.  This library is free
@@ -40,6 +40,7 @@
 #define _GLIBCXX_IOMANIP 1
 
 #pragma GCC system_header
+#pragma GCC visibility push(default)
 
 #include <bits/c++config.h>
 #include <istream>
@@ -296,4 +297,6 @@
 #endif
 } // namespace std
 
+#pragma GCC visibility pop
+
 #endif /* _GLIBCXX_IOMANIP */
--- gcc/libstdc++-v3/include/std/std_ios.h.vis1~	2005-10-31 21:16:10.000000000 +0100
+++ gcc/libstdc++-v3/include/std/std_ios.h	2005-11-03 10:28:34.000000000 +0100
@@ -1,6 +1,7 @@
 // Iostreams base classes -*- C++ -*-
 
-// Copyright (C) 1997, 1998, 1999, 2001, 2002 Free Software Foundation, Inc.
+// Copyright (C) 1997, 1998, 1999, 2001, 2002, 2005
+// Free Software Foundation, Inc.
 //
 // This file is part of the GNU ISO C++ Library.  This library is free
 // software; you can redistribute it and/or modify it under the
@@ -39,6 +40,7 @@
 #define _GLIBCXX_IOS 1
 
 #pragma GCC system_header
+#pragma GCC visibility push(default)
 
 #include <iosfwd>
 #include <exception> 		// For ios_base::failure
@@ -49,4 +51,6 @@
 #include <streambuf> 
 #include <bits/basic_ios.h>
 
+#pragma GCC visibility pop
+
 #endif	/* _GLIBCXX_IOS */
--- gcc/libstdc++-v3/include/std/std_iosfwd.h.vis1~	2005-10-31 21:16:10.000000000 +0100
+++ gcc/libstdc++-v3/include/std/std_iosfwd.h	2005-11-03 10:28:34.000000000 +0100
@@ -1,6 +1,6 @@
 // Forwarding declarations -*- C++ -*-
 
-// Copyright (C) 1997, 1998, 1999, 2001, 2002, 2003
+// Copyright (C) 1997, 1998, 1999, 2001, 2002, 2003, 2005
 // Free Software Foundation, Inc.
 //
 // This file is part of the GNU ISO C++ Library.  This library is free
@@ -40,6 +40,7 @@
 #define _GLIBCXX_IOSFWD 1
 
 #pragma GCC system_header
+#pragma GCC visibility push(default)
 
 #include <bits/c++config.h>
 #include <bits/c++locale.h> 
@@ -164,4 +165,6 @@
   /** @}  */
 } // namespace std
 
+#pragma GCC visibility pop
+
 #endif /* _GLIBCXX_IOSFWD */
--- gcc/libstdc++-v3/include/std/std_iostream.h.vis1~	2005-10-31 21:16:10.000000000 +0100
+++ gcc/libstdc++-v3/include/std/std_iostream.h	2005-11-03 10:28:34.000000000 +0100
@@ -1,6 +1,7 @@
 // Standard iostream objects -*- C++ -*-
 
-// Copyright (C) 1997, 1998, 1999, 2001, 2002 Free Software Foundation, Inc.
+// Copyright (C) 1997, 1998, 1999, 2001, 2002, 2005
+// Free Software Foundation, Inc.
 //
 // This file is part of the GNU ISO C++ Library.  This library is free
 // software; you can redistribute it and/or modify it under the
@@ -39,6 +40,7 @@
 #define _GLIBCXX_IOSTREAM 1
 
 #pragma GCC system_header
+#pragma GCC visibility push(default)
 
 #include <bits/c++config.h>
 #include <ostream>
@@ -76,4 +78,6 @@
   static ios_base::Init __ioinit;
 } // namespace std
 
+#pragma GCC visibility pop
+
 #endif /* _GLIBCXX_IOSTREAM */
--- gcc/libstdc++-v3/include/std/std_istream.h.vis1~	2005-10-31 21:16:10.000000000 +0100
+++ gcc/libstdc++-v3/include/std/std_istream.h	2005-11-03 10:28:35.000000000 +0100
@@ -40,6 +40,7 @@
 #define _GLIBCXX_ISTREAM 1
 
 #pragma GCC system_header
+#pragma GCC visibility push(default)
 
 #include <ios>
 #include <limits> // For numeric_limits
@@ -819,4 +820,6 @@
 # include <bits/istream.tcc>
 #endif
 
+#pragma GCC visibility pop
+
 #endif	/* _GLIBCXX_ISTREAM */
--- gcc/libstdc++-v3/include/std/std_iterator.h.vis1~	2005-10-31 21:16:10.000000000 +0100
+++ gcc/libstdc++-v3/include/std/std_iterator.h	2005-11-03 10:28:35.000000000 +0100
@@ -1,6 +1,6 @@
 // <iterator> -*- C++ -*-
 
-// Copyright (C) 2001, 2002 Free Software Foundation, Inc.
+// Copyright (C) 2001, 2002, 2005 Free Software Foundation, Inc.
 //
 // This file is part of the GNU ISO C++ Library.  This library is free
 // software; you can redistribute it and/or modify it under the
@@ -61,6 +61,7 @@
 #define _GLIBCXX_ITERATOR 1
 
 #pragma GCC system_header
+#pragma GCC visibility push(default)
 
 #include <bits/c++config.h>
 #include <cstddef>
@@ -72,4 +73,6 @@
 #include <bits/stream_iterator.h>
 #include <bits/streambuf_iterator.h>
 
+#pragma GCC visibility pop
+
 #endif /* _GLIBCXX_ITERATOR */
--- gcc/libstdc++-v3/include/std/std_limits.h.vis1~	2005-10-31 21:16:10.000000000 +0100
+++ gcc/libstdc++-v3/include/std/std_limits.h	2005-11-03 10:28:35.000000000 +0100
@@ -1,6 +1,7 @@
 // The template and inlines for the -*- C++ -*- numeric_limits classes.
 
-// Copyright (C) 1999, 2000, 2001, 2002, 2003 Free Software Foundation, Inc.
+// Copyright (C) 1999, 2000, 2001, 2002, 2003, 2005
+// Free Software Foundation, Inc.
 //
 // This file is part of the GNU ISO C++ Library.  This library is free
 // software; you can redistribute it and/or modify it under the
@@ -43,6 +44,7 @@
 #define _GLIBCXX_NUMERIC_LIMITS 1
 
 #pragma GCC system_header
+#pragma GCC visibility push(default)
 
 #include <bits/c++config.h>
 
@@ -1156,4 +1158,6 @@
 #undef __glibcxx_digits
 #undef __glibcxx_digits10
 
+#pragma GCC visibility pop
+
 #endif // _GLIBCXX_NUMERIC_LIMITS
--- gcc/libstdc++-v3/include/std/std_list.h.vis1~	2005-10-31 21:16:10.000000000 +0100
+++ gcc/libstdc++-v3/include/std/std_list.h	2005-11-03 10:28:35.000000000 +0100
@@ -1,6 +1,6 @@
 // <list> -*- C++ -*-
 
-// Copyright (C) 2001, 2002, 2003 Free Software Foundation, Inc.
+// Copyright (C) 2001, 2002, 2003, 2005 Free Software Foundation, Inc.
 //
 // This file is part of the GNU ISO C++ Library.  This library is free
 // software; you can redistribute it and/or modify it under the
@@ -61,6 +61,7 @@
 #define _GLIBCXX_LIST 1
 
 #pragma GCC system_header
+#pragma GCC visibility push(default)
 
 #include <bits/functexcept.h>
 #include <bits/stl_algobase.h>
@@ -77,5 +78,7 @@
 # include <debug/list>
 #endif
 
+#pragma GCC visibility pop
+
 #endif /* _GLIBCXX_LIST */
 
--- gcc/libstdc++-v3/include/std/std_locale.h.vis1~	2005-10-31 21:16:10.000000000 +0100
+++ gcc/libstdc++-v3/include/std/std_locale.h	2005-11-03 10:28:35.000000000 +0100
@@ -1,6 +1,7 @@
 // Locale support -*- C++ -*-
 
-// Copyright (C) 1997, 1998, 1999, 2002, 2003 Free Software Foundation, Inc.
+// Copyright (C) 1997, 1998, 1999, 2002, 2003, 2005
+// Free Software Foundation, Inc.
 //
 // This file is part of the GNU ISO C++ Library.  This library is free
 // software; you can redistribute it and/or modify it under the
@@ -39,10 +40,13 @@
 #define _GLIBCXX_LOCALE	1
 
 #pragma GCC system_header
+#pragma GCC visibility push(default)
 
 #include <bits/localefwd.h>
 #include <bits/locale_classes.h>
 #include <bits/locale_facets.h>
 #include <bits/locale_facets.tcc>
 
+#pragma GCC visibility pop
+
 #endif /* _GLIBCXX_LOCALE */
--- gcc/libstdc++-v3/include/std/std_map.h.vis1~	2005-10-31 21:16:10.000000000 +0100
+++ gcc/libstdc++-v3/include/std/std_map.h	2005-11-03 10:28:35.000000000 +0100
@@ -1,6 +1,6 @@
 // <map> -*- C++ -*-
 
-// Copyright (C) 2001, 2002 Free Software Foundation, Inc.
+// Copyright (C) 2001, 2002, 2005 Free Software Foundation, Inc.
 //
 // This file is part of the GNU ISO C++ Library.  This library is free
 // software; you can redistribute it and/or modify it under the
@@ -61,6 +61,7 @@
 #define _GLIBCXX_MAP 1
 
 #pragma GCC system_header
+#pragma GCC visibility push(default)
 
 #include <bits/stl_tree.h>
 #include <bits/stl_map.h>
@@ -70,4 +71,6 @@
 # include <debug/map>
 #endif
 
+#pragma GCC visibility pop
+
 #endif /* _GLIBCXX_MAP */
--- gcc/libstdc++-v3/include/std/std_memory.h.vis1~	2005-10-31 21:16:10.000000000 +0100
+++ gcc/libstdc++-v3/include/std/std_memory.h	2005-11-03 10:28:35.000000000 +0100
@@ -1,6 +1,6 @@
 // <memory> -*- C++ -*-
 
-// Copyright (C) 2001, 2002, 2004 Free Software Foundation, Inc.
+// Copyright (C) 2001, 2002, 2004, 2005 Free Software Foundation, Inc.
 //
 // This file is part of the GNU ISO C++ Library.  This library is free
 // software; you can redistribute it and/or modify it under the
@@ -49,6 +49,7 @@
 #define _GLIBCXX_MEMORY 1
 
 #pragma GCC system_header
+#pragma GCC visibility push(default)
 
 #include <bits/stl_algobase.h>
 #include <bits/allocator.h>
@@ -370,4 +371,6 @@
   };
 } // namespace std
 
+#pragma GCC visibility pop
+
 #endif /* _GLIBCXX_MEMORY */
--- gcc/libstdc++-v3/include/std/std_numeric.h.vis1~	2005-10-31 21:16:10.000000000 +0100
+++ gcc/libstdc++-v3/include/std/std_numeric.h	2005-11-03 10:28:35.000000000 +0100
@@ -1,6 +1,6 @@
 // <numeric> -*- C++ -*-
 
-// Copyright (C) 2001, 2002 Free Software Foundation, Inc.
+// Copyright (C) 2001, 2002, 2005 Free Software Foundation, Inc.
 //
 // This file is part of the GNU ISO C++ Library.  This library is free
 // software; you can redistribute it and/or modify it under the
@@ -61,6 +61,7 @@
 #define _GLIBCXX_NUMERIC 1
 
 #pragma GCC system_header
+#pragma GCC visibility push(default)
 
 #include <bits/c++config.h>
 #include <cstddef>
@@ -68,4 +69,6 @@
 #include <bits/stl_function.h>
 #include <bits/stl_numeric.h>
 
+#pragma GCC visibility pop
+
 #endif /* _GLIBCXX_NUMERIC */
--- gcc/libstdc++-v3/include/std/std_ostream.h.vis1~	2005-10-31 21:16:10.000000000 +0100
+++ gcc/libstdc++-v3/include/std/std_ostream.h	2005-11-03 10:28:35.000000000 +0100
@@ -1,6 +1,6 @@
 // Output streams -*- C++ -*-
 
-// Copyright (C) 1997, 1998, 1999, 2000, 2001, 2002, 2003
+// Copyright (C) 1997, 1998, 1999, 2000, 2001, 2002, 2003, 2005
 // Free Software Foundation, Inc.
 //
 // This file is part of the GNU ISO C++ Library.  This library is free
@@ -40,6 +40,7 @@
 #define _GLIBCXX_OSTREAM 1
 
 #pragma GCC system_header
+#pragma GCC visibility push(default)
 
 #include <ios>
 
@@ -525,4 +526,6 @@
 # include <bits/ostream.tcc>
 #endif
 
+#pragma GCC visibility pop
+
 #endif	/* _GLIBCXX_OSTREAM */
--- gcc/libstdc++-v3/include/std/std_queue.h.vis1~	2005-10-31 21:16:10.000000000 +0100
+++ gcc/libstdc++-v3/include/std/std_queue.h	2005-11-03 10:28:35.000000000 +0100
@@ -1,6 +1,6 @@
 // <queue> -*- C++ -*-
 
-// Copyright (C) 2001, 2002, 2003 Free Software Foundation, Inc.
+// Copyright (C) 2001, 2002, 2003, 2005 Free Software Foundation, Inc.
 //
 // This file is part of the GNU ISO C++ Library.  This library is free
 // software; you can redistribute it and/or modify it under the
@@ -61,6 +61,7 @@
 #define _GLIBCXX_QUEUE 1
 
 #pragma GCC system_header
+#pragma GCC visibility push(default)
 
 #include <bits/c++config.h>
 #include <bits/functexcept.h>
@@ -74,4 +75,6 @@
 #include <vector>
 #include <bits/stl_queue.h>
 
+#pragma GCC visibility pop
+
 #endif /* _GLIBCXX_QUEUE */
--- gcc/libstdc++-v3/include/std/std_set.h.vis1~	2005-10-31 21:16:10.000000000 +0100
+++ gcc/libstdc++-v3/include/std/std_set.h	2005-11-03 10:28:35.000000000 +0100
@@ -1,6 +1,6 @@
 // <set> -*- C++ -*-
 
-// Copyright (C) 2001, 2002 Free Software Foundation, Inc.
+// Copyright (C) 2001, 2002, 2005 Free Software Foundation, Inc.
 //
 // This file is part of the GNU ISO C++ Library.  This library is free
 // software; you can redistribute it and/or modify it under the
@@ -61,7 +61,8 @@
 #define _GLIBCXX_SET 1
 
 #pragma GCC system_header
-
+#pragma GCC visibility push(default)
+ 
 #include <bits/stl_tree.h>
 #include <bits/stl_set.h>
 #include <bits/stl_multiset.h>
@@ -70,4 +71,6 @@
 # include <debug/set>
 #endif
 
+#pragma GCC visibility pop
+
 #endif /* _GLIBCXX_SET */
--- gcc/libstdc++-v3/include/std/std_sstream.h.vis1~	2005-10-31 21:16:10.000000000 +0100
+++ gcc/libstdc++-v3/include/std/std_sstream.h	2005-11-03 10:28:36.000000000 +0100
@@ -40,6 +40,7 @@
 #define _GLIBCXX_SSTREAM 1
 
 #pragma GCC system_header
+#pragma GCC visibility push(default)
 
 #include <istream>
 #include <ostream>
@@ -614,4 +615,6 @@
 # include <bits/sstream.tcc>
 #endif
 
+#pragma GCC visibility pop
+
 #endif /* _GLIBCXX_SSTREAM */
--- gcc/libstdc++-v3/include/std/std_stack.h.vis1~	2005-10-31 21:16:10.000000000 +0100
+++ gcc/libstdc++-v3/include/std/std_stack.h	2005-11-03 10:28:36.000000000 +0100
@@ -1,6 +1,6 @@
 // <stack> -*- C++ -*-
 
-// Copyright (C) 2001, 2002, 2003 Free Software Foundation, Inc.
+// Copyright (C) 2001, 2002, 2003, 2005 Free Software Foundation, Inc.
 //
 // This file is part of the GNU ISO C++ Library.  This library is free
 // software; you can redistribute it and/or modify it under the
@@ -61,6 +61,7 @@
 #define _GLIBCXX_STACK 1
 
 #pragma GCC system_header
+#pragma GCC visibility push(default)
 
 #include <bits/stl_algobase.h>
 #include <bits/allocator.h>
@@ -69,4 +70,6 @@
 #include <deque>
 #include <bits/stl_stack.h>
 
+#pragma GCC visibility pop
+
 #endif /* _GLIBCXX_STACK */
--- gcc/libstdc++-v3/include/std/std_stdexcept.h.vis1~	2005-10-31 21:16:10.000000000 +0100
+++ gcc/libstdc++-v3/include/std/std_stdexcept.h	2005-11-03 10:28:36.000000000 +0100
@@ -1,6 +1,6 @@
 // Standard exception classes  -*- C++ -*-
 
-// Copyright (C) 2001, 2002 Free Software Foundation, Inc.
+// Copyright (C) 2001, 2002, 2005 Free Software Foundation, Inc.
 //
 // This file is part of the GNU ISO C++ Library.  This library is free
 // software; you can redistribute it and/or modify it under the
@@ -39,6 +39,7 @@
 #define _GLIBCXX_STDEXCEPT 1
 
 #pragma GCC system_header
+#pragma GCC visibility push(default)
 
 #include <exception>
 #include <string>
@@ -144,4 +145,6 @@
   };
 } // namespace std
 
+#pragma GCC visibility pop
+
 #endif /* _GLIBCXX_STDEXCEPT */
--- gcc/libstdc++-v3/include/std/std_streambuf.h.vis1~	2005-10-31 21:16:10.000000000 +0100
+++ gcc/libstdc++-v3/include/std/std_streambuf.h	2005-11-03 10:28:37.000000000 +0100
@@ -40,6 +40,7 @@
 #define _GLIBXX_STREAMBUF 1
 
 #pragma GCC system_header
+#pragma GCC visibility push(default)
 
 #include <bits/c++config.h>
 #include <iosfwd>
@@ -806,4 +807,6 @@
 # include <bits/streambuf.tcc>
 #endif
 
+#pragma GCC visibility pop
+
 #endif /* _GLIBCXX_STREAMBUF */
--- gcc/libstdc++-v3/include/std/std_string.h.vis1~	2005-10-31 21:16:10.000000000 +0100
+++ gcc/libstdc++-v3/include/std/std_string.h	2005-11-03 10:28:37.000000000 +0100
@@ -40,6 +40,7 @@
 #define _GLIBCXX_STRING	1
 
 #pragma GCC system_header
+#pragma GCC visibility push(default)
 
 #include <bits/c++config.h>
 #include <bits/stringfwd.h>
@@ -56,4 +57,6 @@
 # include <bits/basic_string.tcc> 
 #endif
 
+#pragma GCC visibility pop
+
 #endif /* _GLIBCXX_STRING */
--- gcc/libstdc++-v3/include/std/std_utility.h.vis1~	2005-10-31 21:16:10.000000000 +0100
+++ gcc/libstdc++-v3/include/std/std_utility.h	2005-11-03 10:28:37.000000000 +0100
@@ -1,6 +1,6 @@
 // <utility> -*- C++ -*-
 
-// Copyright (C) 2001, 2002 Free Software Foundation, Inc.
+// Copyright (C) 2001, 2002, 2005 Free Software Foundation, Inc.
 //
 // This file is part of the GNU ISO C++ Library.  This library is free
 // software; you can redistribute it and/or modify it under the
@@ -61,9 +61,12 @@
 #define _GLIBCXX_UTILITY 1
 
 #pragma GCC system_header
+#pragma GCC visibility push(default)
 
 #include <bits/c++config.h>
 #include <bits/stl_relops.h>
 #include <bits/stl_pair.h>
 
+#pragma GCC visibility pop
+
 #endif /* _GLIBCXX_UTILITY */
--- gcc/libstdc++-v3/include/std/std_valarray.h.vis1~	2005-10-31 21:16:10.000000000 +0100
+++ gcc/libstdc++-v3/include/std/std_valarray.h	2005-11-03 10:28:37.000000000 +0100
@@ -1,6 +1,6 @@
 // The template and inlines for the -*- C++ -*- valarray class.
 
-// Copyright (C) 1997, 1998, 1999, 2000, 2001, 2002, 2004
+// Copyright (C) 1997, 1998, 1999, 2000, 2001, 2002, 2004, 2005
 // Free Software Foundation, Inc.
 //
 // This file is part of the GNU ISO C++ Library.  This library is free
@@ -38,6 +38,7 @@
 #define _GLIBCXX_VALARRAY 1
 
 #pragma GCC system_header
+#pragma GCC visibility push(default)
 
 #include <bits/c++config.h>
 #include <cstddef>
@@ -1007,4 +1008,6 @@
 
 } // namespace std
 
+#pragma GCC visibility pop
+
 #endif /* _GLIBCXX_VALARRAY */
--- gcc/libstdc++-v3/include/std/std_vector.h.vis1~	2005-10-31 21:16:10.000000000 +0100
+++ gcc/libstdc++-v3/include/std/std_vector.h	2005-11-03 10:28:37.000000000 +0100
@@ -1,6 +1,6 @@
 // <vector> -*- C++ -*-
 
-// Copyright (C) 2001, 2002, 2003 Free Software Foundation, Inc.
+// Copyright (C) 2001, 2002, 2003, 2005 Free Software Foundation, Inc.
 //
 // This file is part of the GNU ISO C++ Library.  This library is free
 // software; you can redistribute it and/or modify it under the
@@ -61,6 +61,7 @@
 #define _GLIBCXX_VECTOR 1
 
 #pragma GCC system_header
+#pragma GCC visibility push(default)
 
 #include <bits/functexcept.h>
 #include <bits/stl_algobase.h>
@@ -78,5 +79,7 @@
 # include <debug/vector>
 #endif
 
+#pragma GCC visibility pop
+
 #endif /* _GLIBCXX_VECTOR */
 
--- gcc/libstdc++-v3/include/c_std/std_cassert.h.vis1~	2005-10-31 21:16:11.000000000 +0100
+++ gcc/libstdc++-v3/include/c_std/std_cassert.h	2005-11-03 10:27:12.000000000 +0100
@@ -45,5 +45,8 @@
 // No include guards on this header...
 
 #pragma GCC system_header
+#pragma GCC visibility push(default)
 
 #include <assert.h>
+
+#pragma GCC visibility pop
--- gcc/libstdc++-v3/include/c_std/std_cctype.h.vis1~	2005-10-31 21:16:11.000000000 +0100
+++ gcc/libstdc++-v3/include/c_std/std_cctype.h	2005-11-03 10:27:12.000000000 +0100
@@ -46,6 +46,7 @@
 #define _GLIBCXX_CCTYPE 1
 
 #pragma GCC system_header
+#pragma GCC visibility push(default)
 
 #include <bits/c++config.h>
 #include <ctype.h>
@@ -82,4 +83,6 @@
   using ::toupper;
 }
 
+#pragma GCC visibility pop
+
 #endif
--- gcc/libstdc++-v3/include/c_std/std_cerrno.h.vis1~	2005-10-31 21:16:11.000000000 +0100
+++ gcc/libstdc++-v3/include/c_std/std_cerrno.h	2005-11-03 10:27:12.000000000 +0100
@@ -46,6 +46,7 @@
 #define _GLIBCXX_CERRNO 1
 
 #pragma GCC system_header
+#pragma GCC visibility push(default)
 
 #include <errno.h>
 
@@ -54,4 +55,6 @@
 #define errno errno
 #endif
 
+#pragma GCC visibility pop
+
 #endif
--- gcc/libstdc++-v3/include/c_std/std_cfloat.h.vis1~	2005-10-31 21:16:11.000000000 +0100
+++ gcc/libstdc++-v3/include/c_std/std_cfloat.h	2005-11-03 10:27:12.000000000 +0100
@@ -46,7 +46,10 @@
 #define _GLIBCXX_CFLOAT 1
 
 #pragma GCC system_header
+#pragma GCC visibility push(default)
 
 #include <float.h>
 
+#pragma GCC visibility pop
+
 #endif
--- gcc/libstdc++-v3/include/c_std/std_climits.h.vis1~	2005-10-31 21:16:11.000000000 +0100
+++ gcc/libstdc++-v3/include/c_std/std_climits.h	2005-11-03 10:27:12.000000000 +0100
@@ -46,7 +46,9 @@
 #define _GLIBCXX_CLIMITS 1
 
 #pragma GCC system_header
+#pragma GCC visibility push(default)
 
 #include <limits.h>
 
+#pragma GCC visibility pop
 #endif
--- gcc/libstdc++-v3/include/c_std/std_clocale.h.vis1~	2005-10-31 21:16:11.000000000 +0100
+++ gcc/libstdc++-v3/include/c_std/std_clocale.h	2005-11-03 10:27:12.000000000 +0100
@@ -46,6 +46,7 @@
 #define _GLIBCXX_CLOCALE 1
 
 #pragma GCC system_header
+#pragma GCC visibility push(default)
 
 #include <locale.h>
 
@@ -60,4 +61,6 @@
   using ::localeconv;
 }
 
+#pragma GCC visibility pop
+
 #endif
--- gcc/libstdc++-v3/include/c_std/std_cmath.h.vis1~	2005-10-31 21:16:11.000000000 +0100
+++ gcc/libstdc++-v3/include/c_std/std_cmath.h	2005-11-03 10:27:12.000000000 +0100
@@ -46,6 +46,7 @@
 #define _GLIBCXX_CMATH 1
 
 #pragma GCC system_header
+#pragma GCC visibility push(default)
 
 #include <bits/c++config.h>
 #include <bits/cpp_type_traits.h>
@@ -575,4 +576,6 @@
 # include <bits/cmath.tcc>
 #endif
 
+#pragma GCC visibility pop
+
 #endif
--- gcc/libstdc++-v3/include/c_std/std_csetjmp.h.vis1~	2005-10-31 21:16:11.000000000 +0100
+++ gcc/libstdc++-v3/include/c_std/std_csetjmp.h	2005-11-03 10:27:12.000000000 +0100
@@ -46,6 +46,7 @@
 #define _GLIBCXX_CSETJMP 1
 
 #pragma GCC system_header
+#pragma GCC visibility push(default)
 
 #include <setjmp.h>
 
@@ -63,4 +64,6 @@
   using ::longjmp;
 }
 
+#pragma GCC visibility pop
+
 #endif
--- gcc/libstdc++-v3/include/c_std/std_csignal.h.vis1~	2005-10-31 21:16:11.000000000 +0100
+++ gcc/libstdc++-v3/include/c_std/std_csignal.h	2005-11-03 10:27:12.000000000 +0100
@@ -46,6 +46,7 @@
 #define _GLIBCXX_CSIGNAL 1
 
 #pragma GCC system_header
+#pragma GCC visibility push(default)
 
 #include <signal.h>
 
@@ -59,4 +60,6 @@
   using ::raise;
 }
 
+#pragma GCC visibility pop
+
 #endif
--- gcc/libstdc++-v3/include/c_std/std_cstdarg.h.vis1~	2005-10-31 21:16:11.000000000 +0100
+++ gcc/libstdc++-v3/include/c_std/std_cstdarg.h	2005-11-03 10:27:12.000000000 +0100
@@ -46,6 +46,7 @@
 #define _GLIBCXX_CSTDARG 1
 
 #pragma GCC system_header
+#pragma GCC visibility push(default)
 
 #include <stdarg.h>
 
@@ -59,4 +60,6 @@
   using ::va_list;
 }
 
+#pragma GCC visibility pop
+
 #endif
--- gcc/libstdc++-v3/include/c_std/std_cstddef.h.vis1~	2005-10-31 21:16:11.000000000 +0100
+++ gcc/libstdc++-v3/include/c_std/std_cstddef.h	2005-11-03 10:27:12.000000000 +0100
@@ -46,6 +46,7 @@
 #define _GLIBCXX_CSTDDEF 1
 
 #pragma GCC system_header
+#pragma GCC visibility push(default)
 
 #include <stddef.h>
 
@@ -55,4 +56,6 @@
   using ::size_t;
 }
 
+#pragma GCC visibility pop
+
 #endif
--- gcc/libstdc++-v3/include/c_std/std_cstdio.h.vis1~	2005-10-31 21:16:11.000000000 +0100
+++ gcc/libstdc++-v3/include/c_std/std_cstdio.h	2005-11-03 10:27:12.000000000 +0100
@@ -46,6 +46,7 @@
 #define _GLIBCXX_CSTDIO 1
 
 #pragma GCC system_header
+#pragma GCC visibility push(default)
 
 #include <bits/c++config.h>
 #include <cstddef>
@@ -184,4 +185,6 @@
 }
 #endif
 
+#pragma GCC visibility pop
+
 #endif
--- gcc/libstdc++-v3/include/c_std/std_cstdlib.h.vis1~	2005-10-31 21:16:11.000000000 +0100
+++ gcc/libstdc++-v3/include/c_std/std_cstdlib.h	2005-11-03 10:27:12.000000000 +0100
@@ -46,6 +46,7 @@
 #define _GLIBCXX_CSTDLIB 1
 
 #pragma GCC system_header
+#pragma GCC visibility push(default)
 
 #include <bits/c++config.h>
 #include <cstddef>
@@ -215,4 +216,6 @@
 }
 #endif
 
+#pragma GCC visibility pop
+
 #endif
--- gcc/libstdc++-v3/include/c_std/std_cstring.h.vis1~	2005-10-31 21:16:11.000000000 +0100
+++ gcc/libstdc++-v3/include/c_std/std_cstring.h	2005-11-03 10:27:12.000000000 +0100
@@ -46,6 +46,7 @@
 #define _GLIBCXX_CSTRING 1
 
 #pragma GCC system_header
+#pragma GCC visibility push(default)
 
 #include <cstddef>
 
@@ -126,4 +127,6 @@
   { return __builtin_strstr(const_cast<const char*>(__s1), __s2); }
 }
 
+#pragma GCC visibility pop
+
 #endif
--- gcc/libstdc++-v3/include/c_std/std_ctime.h.vis1~	2005-10-31 21:16:11.000000000 +0100
+++ gcc/libstdc++-v3/include/c_std/std_ctime.h	2005-11-03 10:27:12.000000000 +0100
@@ -46,6 +46,7 @@
 #define _GLIBCXX_CTIME 1
 
 #pragma GCC system_header
+#pragma GCC visibility push(default)
 
 #include <cstddef>
 
@@ -79,4 +80,6 @@
   using ::strftime;
 }
 
+#pragma GCC visibility pop
+
 #endif
--- gcc/libstdc++-v3/include/c_std/std_cwchar.h.vis1~	2005-10-31 21:16:11.000000000 +0100
+++ gcc/libstdc++-v3/include/c_std/std_cwchar.h	2005-11-03 10:27:12.000000000 +0100
@@ -46,6 +46,7 @@
 #define _GLIBCXX_CWCHAR 1
 
 #pragma GCC system_header
+#pragma GCC visibility push(default)
 
 #include <bits/c++config.h>
 #include <cstddef>
@@ -271,4 +272,6 @@
 
 #endif //_GLIBCXX_USE_WCHAR_T
 
+#pragma GCC visibility pop
+
 #endif
--- gcc/libstdc++-v3/include/c_std/std_cwctype.h.vis1~	2005-10-31 21:16:11.000000000 +0100
+++ gcc/libstdc++-v3/include/c_std/std_cwctype.h	2005-11-03 10:27:12.000000000 +0100
@@ -46,6 +46,7 @@
 #define _GLIBCXX_CWCTYPE 1
 
 #pragma GCC system_header
+#pragma GCC visibility push(default)
 
 #include <bits/c++config.h>
 
@@ -106,4 +107,6 @@
 }
 #endif //_GLIBCXX_USE_WCHAR_T
 
+#pragma GCC visibility pop
+
 #endif
--- gcc/libstdc++-v3/testsuite/20_util/memory/auto_ptr/assign_neg.cc.vis1~	2005-10-31 21:19:34.000000000 +0100
+++ gcc/libstdc++-v3/testsuite/20_util/memory/auto_ptr/assign_neg.cc	2005-11-03 10:28:38.000000000 +0100
@@ -46,5 +46,5 @@
   test01();
   return 0;
 }
-// { dg-error "candidates" "" { target *-*-* } 223 } 
-// { dg-error "std::auto_ptr" "" { target *-*-* } 353 } 
+// { dg-error "candidates" "" { target *-*-* } 224 } 
+// { dg-error "std::auto_ptr" "" { target *-*-* } 354 } 
