--- gcc-3.2.2/gcc/ada/Makefile.in.orig	Thu Mar  6 14:19:49 2003
+++ gcc-3.2.2/gcc/ada/Makefile.in	Thu Mar  6 14:23:02 2003
@@ -1514,7 +1514,7 @@
 
 LIBGNAT_OBJS = adaint.o argv.o cio.o cstreams.o errno.o exit.o \
   raise.o sysdep.o io-aux.o init.o cal.o adafinal.o \
-  tracebak.o expect.o ../../prefix.o $(EXTRA_LIBGNAT_OBJS)
+  tracebak.o expect.o $(EXTRA_LIBGNAT_OBJS)
 
 # NOTE ??? - when the -I option for compiling Ada code is made to work,
 #  the library installation will change and there will be a
