*** trunk/gcc/tree-vrp.c	(revision 115200)
--- trunk/gcc/tree-vrp.c	(working copy)
*************** update_value_range (tree var, value_rang
*** 300,307 ****
    /* Update the value range, if necessary.  */
    old_vr = get_value_range (var);
    is_new = old_vr->type != new_vr->type
!            || old_vr->min != new_vr->min
! 	   || old_vr->max != new_vr->max
  	   || (old_vr->equiv == NULL && new_vr->equiv)
  	   || (old_vr->equiv && new_vr->equiv == NULL)
  	   || (!bitmap_equal_p (old_vr->equiv, new_vr->equiv));
--- 300,313 ----
    /* Update the value range, if necessary.  */
    old_vr = get_value_range (var);
    is_new = old_vr->type != new_vr->type
! 	   || (old_vr->min == NULL && new_vr->min)
! 	   || (old_vr->min && new_vr->min == NULL)
! 	   || (old_vr->max == NULL && new_vr->max)
! 	   || (old_vr->max && new_vr->max == NULL)
!            || (old_vr->min && new_vr->min
! 	       && !operand_equal_p (old_vr->min, new_vr->min, 0))
! 	   || (old_vr->max && new_vr->max
! 	       && !operand_equal_p (old_vr->max, new_vr->max, 0))
  	   || (old_vr->equiv == NULL && new_vr->equiv)
  	   || (old_vr->equiv && new_vr->equiv == NULL)
  	   || (!bitmap_equal_p (old_vr->equiv, new_vr->equiv));
