--- gcc-4.0-20050305/gcc/gimplify.c.orig	2005-02-18 20:35:37.000000000 +0100
+++ gcc-4.0-20050305/gcc/gimplify.c	2005-03-11 00:45:51.600048768 +0100
@@ -2841,28 +2841,6 @@
   while (ret != GS_UNHANDLED)
     switch (TREE_CODE (*from_p))
       {
-      case INDIRECT_REF:
-	{
-	  /* If we have code like 
-
-	        *(const A*)(A*)&x
-
-	     where the type of "x" is a (possibly cv-qualified variant
-	     of "A"), treat the entire expression as identical to "x".
-	     This kind of code arises in C++ when an object is bound
-	     to a const reference, and if "x" is a TARGET_EXPR we want
-	     to take advantage of the optimization below.  */
-	  tree t = fold_indirect_ref (*from_p);
-	  if (t != *from_p)
-	    {
-	      *from_p = t;
-	      ret = GS_OK;
-	    }
-	  else
-	    ret = GS_UNHANDLED;
-	  break;
-	}
-
       case TARGET_EXPR:
 	{
 	  /* If we are initializing something from a TARGET_EXPR, strip the
