--- libjava/configure.ac.orig	2012-03-24 09:51:34.000000000 +0100
+++ libjava/configure.ac	2012-03-24 11:44:40.819510707 +0100
@@ -1833,6 +1833,7 @@
 		    [disable symbol versioning for libjava]),
        [case "$enableval" in
          yes) libjava_cv_anon_version_script=yes ;;
+         gnu*) libjava_cv_anon_version_script=yes ;;
          no)  libjava_cv_anon_version_script=no ;;
          *)   AC_MSG_ERROR([Unknown argument to enable/disable symvers]);;
         esac],
--- libjava/configure.orig	2012-03-24 21:48:30.000000000 +0100
+++ libjava/configure	2012-03-24 22:04:51.884786196 +0100
@@ -24570,6 +24570,7 @@
 if test "${enable_symvers+set}" = set; then :
   enableval=$enable_symvers; case "$enableval" in
          yes) libjava_cv_anon_version_script=yes ;;
+         gnu*) libjava_cv_anon_version_script=yes ;;
          no)  libjava_cv_anon_version_script=no ;;
          *)   as_fn_error "Unknown argument to enable/disable symvers" "$LINENO" 5;;
         esac
