PR target/16092 fix backport from gcc 3.4.0
--- gcc-3.3.4/gcc/config/i386/i386.c.orig	2004-05-18 07:07:52.000000000 +0200
+++ gcc-3.3.4/gcc/config/i386/i386.c	2004-06-20 19:17:22.244016950 +0200
@@ -14676,12 +14676,12 @@
   xops[0] = XEXP (DECL_RTL (function), 0);
   if (TARGET_64BIT)
     {
+      xops[0] = XEXP (xops[0], 0);
       if (!flag_pic || (*targetm.binds_local_p) (function))
 	output_asm_insn ("jmp\t%P0", xops);
       else
 	{
-	  tmp = XEXP (xops[0], 0);
-	  tmp = gen_rtx_UNSPEC (Pmode, gen_rtvec (1, tmp), UNSPEC_GOTPCREL);
+	  tmp = gen_rtx_UNSPEC (Pmode, gen_rtvec (1, xops[0]), UNSPEC_GOTPCREL);
 	  tmp = gen_rtx_CONST (Pmode, tmp);
 	  tmp = gen_rtx_MEM (QImode, tmp);
 	  xops[0] = tmp;
