--- trunk/libjava/classpath/configure.ac	2006-07-02 15:32:04.000000000 +0200
+++ trunk/libjava/classpath/configure.ac	2006-07-02 19:18:28.913906000 +0200
@@ -502,7 +502,8 @@
       AC_CHECK_FILE([$EXTRA_QT_INCLUDE_DIR/QWidget],
 	QT_CFLAGS="$QT_CFLAGS -I$EXTRA_QT_INCLUDE_DIR",
 	AC_MSG_WARN([QWidget not found])))
-	AC_CHECK_PROG(MOC, [moc], [moc])
+	QT4DIR=`$PKG_CONFIG --variable=prefix QtGui`
+	AC_CHECK_PROG(MOC, [moc], [$QT4DIR/bin/moc], [], $QT4DIR/bin)
 	AC_CHECK_PROG(MOC, [moc-qt4], [moc-qt4])
     fi
     if test "x$HAVE_QT4" = "xno"; then
--- trunk/libjava/classpath/native/jni/qt-peer/Makefile.am	2007-06-04 01:18:43.000000000 +0200
+++ trunk/libjava/classpath/native/jni/qt-peer/Makefile.am	2008-03-06 16:32:36.000000000 +0100
@@ -1,8 +1,7 @@
 # Qt AWT backend for Classpath
 #
 
-## GCJ LOCAL: don't install this library
-noinst_LTLIBRARIES = libqtpeer.la
+nativeexeclib_LTLIBRARIES = libqtpeer.la
 
 AM_LDFLAGS = @CLASSPATH_MODULE@ @QT_LIBS@
 AM_CPPFLAGS = @CLASSPATH_INCLUDES@
--- trunk/libjava/classpath/native/jni/qt-peer/Makefile.in	2007-08-04 12:53:49.000000000 +0200
+++ trunk/libjava/classpath/native/jni/qt-peer/Makefile.in	2008-03-06 15:24:38.000000000 +0100
@@ -64,7 +64,8 @@
 mkinstalldirs = $(SHELL) $(top_srcdir)/../../mkinstalldirs
 CONFIG_HEADER = $(top_builddir)/include/config.h
 CONFIG_CLEAN_FILES =
-LTLIBRARIES = $(noinst_LTLIBRARIES)
+nativeexeclibLTLIBRARIES_INSTALL = $(INSTALL)
+LTLIBRARIES = $(nativeexeclib_LTLIBRARIES)
 libqtpeer_la_LIBADD =
 am_libqtpeer_la_OBJECTS = componentevent.lo keybindings.lo \
 	mainqtthread.lo mainthreadinterface.lo nativewrapper.lo \
@@ -357,7 +358,7 @@
 target_vendor = @target_vendor@
 toolexeclibdir = @toolexeclibdir@
 vm_classes = @vm_classes@
-noinst_LTLIBRARIES = libqtpeer.la
+nativeexeclib_LTLIBRARIES = libqtpeer.la
 AM_LDFLAGS = @CLASSPATH_MODULE@ @QT_LIBS@
 AM_CPPFLAGS = @CLASSPATH_INCLUDES@
 AM_CXXFLAGS = @QT_CFLAGS@ \
@@ -475,7 +476,7 @@
 	  rm -f "$${dir}/so_locations"; \
 	done
 libqtpeer.la: $(libqtpeer_la_OBJECTS) $(libqtpeer_la_DEPENDENCIES) 
-	$(CXXLINK)  $(libqtpeer_la_LDFLAGS) $(libqtpeer_la_OBJECTS) $(libqtpeer_la_LIBADD) $(LIBS)
+	$(LINK) -rpath $(nativeexeclibdir) $(libqtpeer_la_LDFLAGS) $(libqtpeer_la_OBJECTS) $(libqtpeer_la_LIBADD) $(LIBS)
 
 mostlyclean-compile:
 	-rm -f *.$(OBJEXT)
@@ -686,7 +687,18 @@
 
 install-data-am:
 
-install-exec-am:
+install-nativeexeclibLTLIBRARIES: $(nativeexeclib_LTLIBRARIES)
+	@$(NORMAL_INSTALL)
+	test -z "$(nativeexeclibdir)" || $(mkdir_p) "$(DESTDIR)$(nativeexeclibdir)"
+	@list='$(nativeexeclib_LTLIBRARIES)'; for p in $$list; do \
+	  if test -f $$p; then \
+	    f=$(am__strip_dir) \
+	    echo " $(LIBTOOL) --mode=install $(nativeexeclibLTLIBRARIES_INSTALL) $(INSTALL_STRIP_FLAG) '$$p' '$(DESTDIR)$(nativeexeclibdir)/$$f'"; \
+	    $(LIBTOOL) --mode=install $(nativeexeclibLTLIBRARIES_INSTALL) $(INSTALL_STRIP_FLAG) "$$p" "$(DESTDIR)$(nativeexeclibdir)/$$f"; \
+	  else :; fi; \
+	done
+
+install-exec-am: install-nativeexeclibLTLIBRARIES
 
 install-info: install-info-am
 
@@ -720,6 +732,7 @@
 	distclean-tags distdir dvi dvi-am html html-am info info-am \
 	install install-am install-data install-data-am install-exec \
 	install-exec-am install-info install-info-am install-man \
+	install-nativeexeclibLTLIBRARIES \
 	install-strip installcheck installcheck-am installdirs \
 	maintainer-clean maintainer-clean-generic mostlyclean \
 	mostlyclean-compile mostlyclean-generic mostlyclean-libtool \
