--- gcc/libjava/configure.ac	2006-01-23 18:53:45.000000000 +0100
+++ gcc/libjava/configure.ac	2006-02-26 16:02:13.828163000 +0100
@@ -84,6 +84,21 @@
   *)   AC_MSG_ERROR(bad value ${enableval} for multilib option) ;;
  esac], [multilib=yes])dnl
 
+# Default to --enable-libjava-multilib
+AC_ARG_ENABLE(libjava-multilib,
+  AS_HELP_STRING([--enable-libjava-multilib],
+                 [build many library versions (default)]),
+[case "${enableval}" in
+  yes) multilib=yes ;;
+  no)  multilib=no ;;
+  *)   AC_MSG_ERROR(bad value ${enableval} for libjava-multilib option) ;;
+ esac], [multilib=yes])dnl
+if test "$multilib" = no; then
+  # Reset also --enable-multilib state, as that is what is looked at
+  # by config-ml.in
+  ac_configure_args="$ac_configure_args --disable-multilib"
+fi
+
 # It may not be safe to run linking tests in AC_PROG_CC/AC_PROG_CXX.
 GCC_NO_EXECUTABLES
 
