--- gcc-4.0-20050213/gcc/tree-ssa-loop-ivopts.c.orig	2005-02-10 21:13:20.000000000 +0100
+++ gcc-4.0-20050213/gcc/tree-ssa-loop-ivopts.c	2005-02-20 17:03:35.000000000 +0100
@@ -4877,15 +4877,17 @@
   
   if (may_eliminate_iv (data, use, cand, &compare, &bound))
     {
+      tree var = var_at_stmt (data->current_loop, cand, use->stmt);
+      tree var_type = TREE_TYPE (var);
+
+      bound = fold_convert (var_type, bound);
       op = force_gimple_operand (unshare_expr (bound), &stmts,
 				 true, NULL_TREE);
 
       if (stmts)
 	bsi_insert_before (&bsi, stmts, BSI_SAME_STMT);
 
-      *use->op_p = build2 (compare, boolean_type_node,
-			  var_at_stmt (data->current_loop,
-				       cand, use->stmt), op);
+      *use->op_p = build2 (compare, boolean_type_node, var, op);
       modify_stmt (use->stmt);
       return;
     }
