--- gcc-4.0-20050312/gcc/calls.c.orig	2005-01-26 16:08:50.000000000 +0100
+++ gcc-4.0-20050312/gcc/calls.c	2005-03-18 13:14:41.000000000 +0100
@@ -1994,8 +1994,10 @@
 	    structure_value_addr = expand_expr (return_arg, NULL_RTX,
 						VOIDmode, EXPAND_NORMAL);
 	  }
+#if 0
 	else if (target && MEM_P (target))
 	  structure_value_addr = XEXP (target, 0);
+#endif
 	else
 	  {
 	    /* For variable-sized objects, we must be called with a target
