diff -urN boehm-gc.org/Makefile.am boehm-gc/Makefile.am
--- boehm-gc.org/Makefile.am	Sat Oct 18 12:32:07 2003
+++ boehm-gc/Makefile.am	Sat Oct 18 12:32:55 2003
@@ -39,7 +39,8 @@
 # linuxthread semaphore functions get linked:
 libgcjgc_la_LIBADD = @addobjs@ $(THREADLIBS)
 libgcjgc_la_DEPENDENCIES = @addobjs@
-libgcjgc_la_LDFLAGS = -version-info 1:1:0 -rpath $(toolexeclibdir)
+libgcjgc_la_LDFLAGS = @LIBGCJGC_EXTRA_LDFLAGS@ -version-info 1:1:0 \
+		      -rpath $(toolexeclibdir)
 
 libgcjgc_convenience_la_LIBADD = @addobjs@
 libgcjgc_convenience_la_DEPENDENCIES = @addobjs@
diff -urN boehm-gc.org/Makefile.in boehm-gc/Makefile.in
--- boehm-gc.org/Makefile.in	Thu Oct 16 22:10:48 2003
+++ boehm-gc/Makefile.in	Sat Oct 18 12:32:55 2003
@@ -133,7 +133,8 @@
 # linuxthread semaphore functions get linked:
 libgcjgc_la_LIBADD = @addobjs@ $(THREADLIBS)
 libgcjgc_la_DEPENDENCIES = @addobjs@
-libgcjgc_la_LDFLAGS = -version-info 1:1:0 -rpath $(toolexeclibdir)
+libgcjgc_la_LDFLAGS = @LIBGCJGC_EXTRA_LDFLAGS@ -version-info 1:1:0 \
+		      -rpath $(toolexeclibdir)
 
 libgcjgc_convenience_la_LIBADD = @addobjs@
 libgcjgc_convenience_la_DEPENDENCIES = @addobjs@
diff -urN boehm-gc.org/configure boehm-gc/configure
--- boehm-gc.org/configure	Sat Oct 18 12:32:07 2003
+++ boehm-gc/configure	Sat Oct 18 12:33:48 2003
@@ -2717,11 +2717,57 @@
 fi
 
 
+echo $ac_n "checking for dlopen in -ldl""... $ac_c" 1>&6
+echo "configure:2635: checking for dlopen in -ldl" >&5
+ac_lib_var=`echo dl'_'dlopen | sed 'y%./+-%__p_%'`
+if eval "test \"`echo '$''{'ac_cv_lib_$ac_lib_var'+set}'`\" = set"; then
+  echo $ac_n "(cached) $ac_c" 1>&6
+else
+  ac_save_LIBS="$LIBS"
+LIBS="-ldl  $LIBS"
+cat > conftest.$ac_ext <<EOF
+#line 2643 "configure"
+#include "confdefs.h"
+/* Override any gcc2 internal prototype to avoid an error.  */
+/* We use char because int might match the return type of a gcc2
+    builtin and then its argument prototype would still apply.  */
+char dlopen();
+
+int main() {
+dlopen()
+; return 0; }
+EOF
+if { (eval echo configure:2654: \"$ac_link\") 1>&5; (eval $ac_link) 2>&5; } && test -s conftest${ac_exeext}; then
+  rm -rf conftest*
+  eval "ac_cv_lib_$ac_lib_var=yes"
+else
+  echo "configure: failed program was:" >&5
+  cat conftest.$ac_ext >&5
+  rm -rf conftest*
+  eval "ac_cv_lib_$ac_lib_var=no"
+fi
+rm -f conftest*
+LIBS="$ac_save_LIBS"
+
+fi
+if eval "test \"`echo '$ac_cv_lib_'$ac_lib_var`\" = yes"; then
+  echo "$ac_t""yes" 1>&6
+  
+EXTRA_TEST_LIBS="$EXTRA_TEST_LIBS -ldl"
+LIBGCJGC_EXTRA_LDFLAGS="-ldl"
+
+else
+  echo "$ac_t""no" 1>&6
+fi
+
+
+
 INCLUDES=-I${srcdir}/include
 THREADLIBS=
 case "$THREADS" in
  no | none | single)
     THREADS=none
+    LIBGCJGC_EXTRA_LDFLAGS=
     ;;
  posix | pthreads)
     THREADS=posix
@@ -2806,6 +2852,7 @@
 	;;
      *-*-cygwin*)
 	THREADLIBS=
+	LIBGCJGC_EXTRA_LDFLAGS=
 	;;
     esac
     ;;
@@ -2834,48 +2881,6 @@
 esac
 
 
-echo $ac_n "checking for dlopen in -ldl""... $ac_c" 1>&6
-echo "configure:2839: checking for dlopen in -ldl" >&5
-ac_lib_var=`echo dl'_'dlopen | sed 'y%./+-%__p_%'`
-if eval "test \"`echo '$''{'ac_cv_lib_$ac_lib_var'+set}'`\" = set"; then
-  echo $ac_n "(cached) $ac_c" 1>&6
-else
-  ac_save_LIBS="$LIBS"
-LIBS="-ldl  $LIBS"
-cat > conftest.$ac_ext <<EOF
-#line 2847 "configure"
-#include "confdefs.h"
-/* Override any gcc2 internal prototype to avoid an error.  */
-/* We use char because int might match the return type of a gcc2
-    builtin and then its argument prototype would still apply.  */
-char dlopen();
-
-int main() {
-dlopen()
-; return 0; }
-EOF
-if { (eval echo configure:2858: \"$ac_link\") 1>&5; (eval $ac_link) 2>&5; } && test -s conftest${ac_exeext}; then
-  rm -rf conftest*
-  eval "ac_cv_lib_$ac_lib_var=yes"
-else
-  echo "configure: failed program was:" >&5
-  cat conftest.$ac_ext >&5
-  rm -rf conftest*
-  eval "ac_cv_lib_$ac_lib_var=no"
-fi
-rm -f conftest*
-LIBS="$ac_save_LIBS"
-
-fi
-if eval "test \"`echo '$ac_cv_lib_'$ac_lib_var`\" = yes"; then
-  echo "$ac_t""yes" 1>&6
-  EXTRA_TEST_LIBS="$EXTRA_TEST_LIBS -ldl"
-else
-  echo "$ac_t""no" 1>&6
-fi
-
-
-
 target_all=libgcjgc.la
 
 
@@ -3295,8 +3300,9 @@
 s%@STRIP@%$STRIP%g
 s%@LIBTOOL@%$LIBTOOL%g
 s%@CXXCPP@%$CXXCPP%g
-s%@THREADLIBS@%$THREADLIBS%g
 s%@EXTRA_TEST_LIBS@%$EXTRA_TEST_LIBS%g
+s%@THREADLIBS@%$THREADLIBS%g
+s%@LIBGCJGC_EXTRA_LDFLAGS@%$LIBGCJGC_EXTRA_LDFLAGS%g
 s%@target_all@%$target_all%g
 s%@INCLUDES@%$INCLUDES%g
 s%@CXXINCLUDES@%$CXXINCLUDES%g
diff -urN boehm-gc.org/configure.in boehm-gc/configure.in
--- boehm-gc.org/configure.in	Sat Oct 18 12:32:07 2003
+++ boehm-gc/configure.in	Sat Oct 18 12:32:55 2003
@@ -63,11 +63,18 @@
     esac]
 )
 
+AC_CHECK_LIB(dl, dlopen, [
+EXTRA_TEST_LIBS="$EXTRA_TEST_LIBS -ldl"
+LIBGCJGC_EXTRA_LDFLAGS="-ldl"
+])
+AC_SUBST(EXTRA_TEST_LIBS)
+
 INCLUDES=-I${srcdir}/include
 THREADLIBS=
 case "$THREADS" in
  no | none | single)
     THREADS=none
+    LIBGCJGC_EXTRA_LDFLAGS=
     ;;
  posix | pthreads)
     THREADS=posix
@@ -110,6 +117,7 @@
 	;;
      *-*-cygwin*)
 	THREADLIBS=
+	LIBGCJGC_EXTRA_LDFLAGS=
 	;;
     esac
     ;;
@@ -128,9 +136,7 @@
     ;;
 esac
 AC_SUBST(THREADLIBS)
-
-AC_CHECK_LIB(dl, dlopen, EXTRA_TEST_LIBS="$EXTRA_TEST_LIBS -ldl")
-AC_SUBST(EXTRA_TEST_LIBS)
+AC_SUBST(LIBGCJGC_EXTRA_LDFLAGS)
 
 target_all=libgcjgc.la
 AC_SUBST(target_all)
