Argh, GDB is not ready.

--- gcc/config/i386/i386.h.jj	2002-08-28 11:20:27.000000000 +0200
+++ gcc/config/i386/i386.h	2002-08-28 22:01:26.000000000 +0200
@@ -390,7 +390,7 @@ extern int x86_prefetch_sse;
 #define TARGET_64BIT_DEFAULT 0
 #endif
 
-#define TARGET_DEFAULT MASK_OMIT_LEAF_FRAME_POINTER
+#define TARGET_DEFAULT 0
 
 /* Which processor to schedule for. The cpu attribute defines a list that
    mirrors this list, so changes to i386.md must be made at the same time.  */
