--- boehm-gc/configure.in.jj	Tue Dec 18 01:27:56 2001
+++ boehm-gc/configure.in	Fri Jan 25 14:10:17 2002
@@ -57,11 +57,18 @@ AC_ARG_ENABLE(parallel-mark,
     esac]
 )
 
+AC_CHECK_LIB(dl, dlopen, [
+EXTRA_TEST_LIBS="$EXTRA_TEST_LIBS -ldl"
+LIBGCJGC_EXTRA_LDFLAGS="-ldl"
+])
+AC_SUBST(EXTRA_TEST_LIBS)
+
 INCLUDES=-I${srcdir}/include
 THREADLIBS=
 case "$THREADS" in
  no | none | single)
     THREADS=none
+    LIBGCJGC_EXTRA_LDFLAGS=
     ;;
  posix | pthreads)
     THREADS=posix
@@ -105,6 +112,7 @@ case "$THREADS" in
 	;;
      *-*-cygwin*)
 	THREADLIBS=
+	LIBGCJGC_EXTRA_LDFLAGS=
 	;;
     esac
     ;;
@@ -116,9 +124,7 @@ case "$THREADS" in
     ;;
 esac
 AC_SUBST(THREADLIBS)
-
-AC_CHECK_LIB(dl, dlopen, EXTRA_TEST_LIBS="$EXTRA_TEST_LIBS -ldl")
-AC_SUBST(EXTRA_TEST_LIBS)
+AC_SUBST(LIBGCJGC_EXTRA_LDFLAGS)
 
 target_all=libgcjgc.la
 AC_SUBST(target_all)
--- boehm-gc/configure.jj	Thu Jan  3 14:25:13 2002
+++ boehm-gc/configure	Fri Jan 25 14:10:36 2002
@@ -2647,11 +2647,57 @@ if test "${enable_parallel_mark+set}" = 
 fi
 
 
+echo $ac_n "checking for dlopen in -ldl""... $ac_c" 1>&6
+echo "configure:2635: checking for dlopen in -ldl" >&5
+ac_lib_var=`echo dl'_'dlopen | sed 'y%./+-%__p_%'`
+if eval "test \"`echo '$''{'ac_cv_lib_$ac_lib_var'+set}'`\" = set"; then
+  echo $ac_n "(cached) $ac_c" 1>&6
+else
+  ac_save_LIBS="$LIBS"
+LIBS="-ldl  $LIBS"
+cat > conftest.$ac_ext <<EOF
+#line 2643 "configure"
+#include "confdefs.h"
+/* Override any gcc2 internal prototype to avoid an error.  */
+/* We use char because int might match the return type of a gcc2
+    builtin and then its argument prototype would still apply.  */
+char dlopen();
+
+int main() {
+dlopen()
+; return 0; }
+EOF
+if { (eval echo configure:2654: \"$ac_link\") 1>&5; (eval $ac_link) 2>&5; } && test -s conftest${ac_exeext}; then
+  rm -rf conftest*
+  eval "ac_cv_lib_$ac_lib_var=yes"
+else
+  echo "configure: failed program was:" >&5
+  cat conftest.$ac_ext >&5
+  rm -rf conftest*
+  eval "ac_cv_lib_$ac_lib_var=no"
+fi
+rm -f conftest*
+LIBS="$ac_save_LIBS"
+
+fi
+if eval "test \"`echo '$ac_cv_lib_'$ac_lib_var`\" = yes"; then
+  echo "$ac_t""yes" 1>&6
+  
+EXTRA_TEST_LIBS="$EXTRA_TEST_LIBS -ldl"
+LIBGCJGC_EXTRA_LDFLAGS="-ldl"
+
+else
+  echo "$ac_t""no" 1>&6
+fi
+
+
+
 INCLUDES=-I${srcdir}/include
 THREADLIBS=
 case "$THREADS" in
  no | none | single)
     THREADS=none
+    LIBGCJGC_EXTRA_LDFLAGS=
     ;;
  posix | pthreads)
     THREADS=posix
@@ -2736,6 +2782,7 @@ EOF
 	;;
      *-*-cygwin*)
 	THREADLIBS=
+	LIBGCJGC_EXTRA_LDFLAGS=
 	;;
     esac
     ;;
@@ -2758,48 +2805,6 @@ EOF
 esac
 
 
-echo $ac_n "checking for dlopen in -ldl""... $ac_c" 1>&6
-echo "configure:2850: checking for dlopen in -ldl" >&5
-ac_lib_var=`echo dl'_'dlopen | sed 'y%./+-%__p_%'`
-if eval "test \"`echo '$''{'ac_cv_lib_$ac_lib_var'+set}'`\" = set"; then
-  echo $ac_n "(cached) $ac_c" 1>&6
-else
-  ac_save_LIBS="$LIBS"
-LIBS="-ldl  $LIBS"
-cat > conftest.$ac_ext <<EOF
-#line 2858 "configure"
-#include "confdefs.h"
-/* Override any gcc2 internal prototype to avoid an error.  */
-/* We use char because int might match the return type of a gcc2
-    builtin and then its argument prototype would still apply.  */
-char dlopen();
-
-int main() {
-dlopen()
-; return 0; }
-EOF
-if { (eval echo configure:2869: \"$ac_link\") 1>&5; (eval $ac_link) 2>&5; } && test -s conftest${ac_exeext}; then
-  rm -rf conftest*
-  eval "ac_cv_lib_$ac_lib_var=yes"
-else
-  echo "configure: failed program was:" >&5
-  cat conftest.$ac_ext >&5
-  rm -rf conftest*
-  eval "ac_cv_lib_$ac_lib_var=no"
-fi
-rm -f conftest*
-LIBS="$ac_save_LIBS"
-
-fi
-if eval "test \"`echo '$ac_cv_lib_'$ac_lib_var`\" = yes"; then
-  echo "$ac_t""yes" 1>&6
-  EXTRA_TEST_LIBS="$EXTRA_TEST_LIBS -ldl"
-else
-  echo "$ac_t""no" 1>&6
-fi
-
-
-
 target_all=libgcjgc.la
 
 
@@ -3215,8 +3220,9 @@ s%@OBJEXT@%$OBJEXT%g
 s%@STRIP@%$STRIP%g
 s%@LIBTOOL@%$LIBTOOL%g
 s%@CXXCPP@%$CXXCPP%g
-s%@THREADLIBS@%$THREADLIBS%g
 s%@EXTRA_TEST_LIBS@%$EXTRA_TEST_LIBS%g
+s%@THREADLIBS@%$THREADLIBS%g
+s%@LIBGCJGC_EXTRA_LDFLAGS@%$LIBGCJGC_EXTRA_LDFLAGS%g
 s%@target_all@%$target_all%g
 s%@INCLUDES@%$INCLUDES%g
 s%@CXXINCLUDES@%$CXXINCLUDES%g
--- boehm-gc/Makefile.am.jj	Mon Oct 22 11:06:57 2001
+++ boehm-gc/Makefile.am	Fri Jan 25 14:11:35 2002
@@ -37,7 +37,8 @@ solaris_pthreads.c solaris_threads.c spe
 # linuxthread semaphore functions get linked:
 libgcjgc_la_LIBADD = @addobjs@ $(THREADLIBS)
 libgcjgc_la_DEPENDENCIES = @addobjs@
-libgcjgc_la_LDFLAGS = -version-info 1:1:0 -rpath $(toolexeclibdir)
+libgcjgc_la_LDFLAGS = @LIBGCJGC_EXTRA_LDFLAGS@ -version-info 1:1:0 \
+		      -rpath $(toolexeclibdir)
 
 EXTRA_libgcjgc_la_SOURCES = alpha_mach_dep.s \
 mips_sgi_mach_dep.s mips_ultrix_mach_dep.s powerpc_macosx_mach_dep.s \
--- boehm-gc/Makefile.in.jj	Thu Jan  3 14:25:13 2002
+++ boehm-gc/Makefile.in	Fri Jan 25 14:12:03 2002
@@ -118,7 +118,8 @@ libgcjgc_la_SOURCES = allchblk.c alloc.c
 # linuxthread semaphore functions get linked:
 libgcjgc_la_LIBADD = @addobjs@ $(THREADLIBS)
 libgcjgc_la_DEPENDENCIES = @addobjs@
-libgcjgc_la_LDFLAGS = -version-info 1:1:0 -rpath $(toolexeclibdir)
+libgcjgc_la_LDFLAGS = @LIBGCJGC_EXTRA_LDFLAGS@ -version-info 1:1:0 \
+		      -rpath $(toolexeclibdir)
 
 EXTRA_libgcjgc_la_SOURCES = alpha_mach_dep.s mips_sgi_mach_dep.s mips_ultrix_mach_dep.s powerpc_macosx_mach_dep.s rs6000_mach_dep.s sparc_mach_dep.s sparc_netbsd_mach_dep.s sparc_sunos4_mach_dep.s ia64_save_regs_in_stack.s
 
