--- libjava/Makefile.am.jj	Tue Jun  4 20:41:44 2002
+++ libjava/Makefile.am	Wed Jun  5 15:20:30 2002
@@ -111,6 +111,8 @@ INCLUDES = -I$(top_srcdir) -Iinclude -I$
 
 ## ################################################################
 
+rpath_def = $(shell if test "$(toolexeclibdir)" != /usr/lib; then echo -rpath $(toolexeclibdir); fi)
+
 ##
 ## How to build libgcj.a and libgcj.jar
 ##
@@ -410,7 +412,7 @@ EXTRA_jv_convert_SOURCES = $(convert_sou
 ## need this because we are explicitly using libtool to link using the
 ## `.la' file.
 jv_convert_LDFLAGS = --main=gnu.gcj.convert.Convert \
-	-rpath $(toolexeclibdir) -shared-libgcc $(THREADLDFLAGS)
+	$(rpath_def) -shared-libgcc $(THREADLDFLAGS)
 jv_convert_LINK = $(GCJLINK)
 ## We don't explicitly link in the libraries we need; libgcj.la brings
 ## in all dependencies.  We need the -L so that gcj can find libgcj
@@ -429,7 +431,7 @@ gij_SOURCES = gij.cc
 ## We need -nodefaultlibs because we want to avoid gcj's `-lgcj'.  We
 ## need this because we are explicitly using libtool to link using the
 ## `.la' file.
-gij_LDFLAGS = -rpath $(toolexeclibdir) -shared-libgcc $(THREADLDFLAGS)
+gij_LDFLAGS = $(rpath_def) -shared-libgcc $(THREADLDFLAGS)
 gij_LINK = $(GCJLINK)
 ## See jv_convert_LDADD.
 gij_LDADD = -L$(here)/.libs libgcj.la
@@ -441,7 +443,7 @@ rmic_SOURCES =
 ## This is a dummy definition.
 EXTRA_rmic_SOURCES = $(rmi_java_source_files)
 rmic_LDFLAGS = --main=gnu.java.rmi.rmic.RMIC \
-	-rpath $(toolexeclibdir) -shared-libgcc $(THREADLDFLAGS)
+	$(rpath_def) -shared-libgcc $(THREADLDFLAGS)
 rmic_LINK = $(GCJLINK)
 ## See jv_convert_LDADD.
 rmic_LDADD = -L$(here)/.libs libgcj.la
@@ -453,7 +455,7 @@ rmiregistry_SOURCES =
 ## This is a dummy definition.
 EXTRA_rmiregistry_SOURCES = $(rmi_java_source_files)
 rmiregistry_LDFLAGS = --main=gnu.java.rmi.registry.RegistryImpl \
-	-rpath $(toolexeclibdir) -shared-libgcc $(THREADLDFLAGS)
+	$(rpath_def) -shared-libgcc $(THREADLDFLAGS)
 rmiregistry_LINK = $(GCJLINK)
 ## See jv_convert_LDADD.
 rmiregistry_LDADD = -L$(here)/.libs libgcj.la
--- libjava/Makefile.in.jj	Tue Jun  4 23:25:56 2002
+++ libjava/Makefile.in	Wed Jun  5 15:21:59 2002
@@ -184,6 +184,7 @@ INCLUDES = -I$(top_srcdir) -Iinclude -I$
 	$(GCINCS) $(THREADINCS) $(INCLTDL) \
 	$(GCC_UNWIND_INCLUDE) $(ZINCS) $(LIBFFIINCS)
 
+rpath_def = $(shell if test "$(toolexeclibdir)" != /usr/lib; then echo -rpath $(toolexeclibdir); fi)
 
 nat_files = $(nat_source_files:.cc=.lo)
 x_nat_files = $(x_nat_source_files:.cc=.lo)
@@ -263,7 +264,7 @@ CONVERT_DIR = gnu/gcj/convert
 jv_convert_SOURCES = 
 EXTRA_jv_convert_SOURCES = $(convert_source_files)
 jv_convert_LDFLAGS = --main=gnu.gcj.convert.Convert \
-	-rpath $(toolexeclibdir) -shared-libgcc $(THREADLDFLAGS)
+	$(rpath_def) -shared-libgcc $(THREADLDFLAGS)
 
 jv_convert_LINK = $(GCJLINK)
 jv_convert_LDADD = -L$(here)/.libs libgcj.la
@@ -272,7 +273,7 @@ jv_convert_DEPENDENCIES = $(convert_sour
 
 
 gij_SOURCES = gij.cc
-gij_LDFLAGS = -rpath $(toolexeclibdir) -shared-libgcc $(THREADLDFLAGS)
+gij_LDFLAGS = $(rpath_def) -shared-libgcc $(THREADLDFLAGS)
 gij_LINK = $(GCJLINK)
 gij_LDADD = -L$(here)/.libs libgcj.la
 gij_DEPENDENCIES = libgcj.la libgcj.spec
@@ -280,7 +281,7 @@ gij_DEPENDENCIES = libgcj.la libgcj.spec
 rmic_SOURCES = 
 EXTRA_rmic_SOURCES = $(rmi_java_source_files)
 rmic_LDFLAGS = --main=gnu.java.rmi.rmic.RMIC \
-	-rpath $(toolexeclibdir) -shared-libgcc $(THREADLDFLAGS)
+	$(rpath_def) -shared-libgcc $(THREADLDFLAGS)
 
 rmic_LINK = $(GCJLINK)
 rmic_LDADD = -L$(here)/.libs libgcj.la
@@ -289,7 +290,7 @@ rmic_DEPENDENCIES = libgcj.la libgcj.spe
 rmiregistry_SOURCES = 
 EXTRA_rmiregistry_SOURCES = $(rmi_java_source_files)
 rmiregistry_LDFLAGS = --main=gnu.java.rmi.registry.RegistryImpl \
-	-rpath $(toolexeclibdir) -shared-libgcc $(THREADLDFLAGS)
+	$(rpath_def) -shared-libgcc $(THREADLDFLAGS)
 
 rmiregistry_LINK = $(GCJLINK)
 rmiregistry_LDADD = -L$(here)/.libs libgcj.la
